---
# knit: bookdown::preview_chapter
output: html_document
---

# Ordering of cells according to pseudotime

```{r, echo=FALSE}
library(knitr)
opts_chunk$set(fig.align = "center")
```
```{r, echo=TRUE, message=FALSE, warning=FALSE}
library(TSCAN)
library(M3Drop)
library(monocle)
set.seed(1)
```

In many situations, one is studying a process where cells change
continuously. This includes for example many differentiation processes
taking place during development, where following a stimulus, cells
will change from one cell-type to another. Ideally, we would like to
monitor the expression levels of an individual cell over
time. Unfortunately, such monitoring is not possible with scRNA-seq
since the cell is lysed (killed) when the RNA is extracted.

Instead, we must sample at specific time-points to obtain snapshots of
the gene expression profiles. However, since some of the cells will
proceed faster along the differentiation than others, the actual time
may not correspond to the developmental progression. Thus, we would
like to order the cells along a one-dimensional trajectory to identify
the progression through state space of which can be interpreted as a
temporal axis (hence the name pseudotime).

In this chapter we will consider two different tools, Monocle and
TSCAN for ordering cells according to their pseudotime development. To
illustrate the methods we will be using a dataset on mouse embryonic
development that was collected by Deng et al. The dataset consists of
268 cells from 10 different time-points of early mouse development.

## TSCAN

First we will try to use all genes to order the cells.
```{r tscan-all-genes}
deng <- readRDS("deng/deng.rds")
cellLabels <- colnames(deng)
procdeng <- TSCAN::preprocess(deng, clusternum=10)
colnames(procdeng) <- 1:268
dengclust <- TSCAN::exprmclust(procdeng, clusternum=10)
TSCAN::plotmclust(dengclust)
dengorderTSCAN <- TSCAN::TSCANorder(dengclust, orderonly=F)
```

## monocle

Unfortunately, Monocle does not work when all the genes are used, so
we must carry out feature selection. First, we use M3D
```{r m3d-select-genes}
m3dGenes <- as.character(
    M3Drop::M3Drop_Differential_Expression(deng, suppress.plot=T)$Gene
)
d <- deng[which(m3dGenes %in% rownames(deng)),]
d <- d[!duplicated(rownames(d)),]
```

```{r monocle-all-genes}
pd <- as.data.frame(colnames(d))
names(pd) <- "timepoint"
pd <- new("AnnotatedDataFrame", data=pd)
fd <- as.data.frame(rownames(d))
names(fd) <- "gene"
fd <- new("AnnotatedDataFrame", data=fd)
colnames(d) <- 1:dim(d)[2]
rownames(d) <- 1:dim(d)[1]
dCellData <- monocle::newCellDataSet(d, phenoData=pd, featureData=fd)
dCellData <- monocle::setOrderingFilter(dCellData, 1:length(m3dGenes))
dCellDataSet <- monocle::reduceDimension(dCellData, pseudo_expr=1)
dCellDataSet <- monocle::orderCells(dCellDataSet, reverse=T)
monocle::plot_spanning_tree(dCellDataSet)
```

As you can see, Monocle has identified 7 different states. This is
probably not such a bad conclusion since we know that three of the
time-points correspond to different stages of the two-cell
embryo. Also, we saw that the three blastocyst stages correspond to
inner cell mass and trophectoderm cells. We can take a closer look at
the clusters
```{r monocle-clusters}
dMonocleClusters <- phenoData(dCellDataSet)$State
```

Exercise: Compare the clusters identified by monocle with the annotated ones. How well do the labels correspond?


Next, we investigate the order of the cells inferred by Monocle
```{r monocle-order}
ts <- as.matrix(phenoData(dCellDataSet)$Pseudotime)
dOrder <- sort.int(ts, index.return=T)$ix
dOrder
```

Now, we repeat the exercise using TSCAN
```{r}
procdeng <- TSCAN::preprocess(d, clusternum=10)
colnames(procdeng) <- 1:268
dengclust <- TSCAN::exprmclust(procdeng, clusternum=10)
TSCAN::plotmclust(dengclust)
dengorderTSCAN <- TSCAN::TSCANorder(dengclust, orderonly=F)
```

Exercise: Use TSCAN to obtain 7 clusters and compare the groups with those obtained from Monocle

## Comparison of the methods

How do the trajectories inferred by TSCAN and Monocle compare?
```{r tscan-monocle-compare}
inds <- dengorderTSCAN$sample_name
plot(
    rev(dengorderTSCAN[,3]), 
    pData(dCellDataSet)[inds,3], 
    col=cellLabels[as.numeric(inds)], 
    pch=20, 
    xlab="TSCAN", 
    ylab="Monocle"
)
```

Exercise: Repeat the exercise using a subset of the genes, e.g. the set of highly variable genes that can be obtained using M3Drop::Brennecke_getVariableGenes

