<!DOCTYPE html>
<html >

<head>

  <meta charset="UTF-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <title>Analysis of single-cell RNA-seq data</title>
  <meta content="text/html; charset=UTF-8" http-equiv="Content-Type">
  <meta name="description" content="Analysis of single-cell RNA-seq data">
  <meta name="generator" content="bookdown <!--bookdown:version--> and GitBook 2.6.7">

<meta name="author" content="Vladimir Kiselev, Tallulah Andrews and Martin Hemberg">

<meta name="date" content="2016-03-23">

<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript"></script>

  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black">
  
  
<!--bookdown:link_prev-->
<!--bookdown:link_next-->

<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<script src="libs/gitbook-2.6.7/js/app.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>


<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>


<!--bookdown:title:start-->
<div id="header">
<h1 class="title">Analysis of single-cell RNA-seq data</h1>
<h4 class="author"><em>Vladimir Kiselev, Tallulah Andrews and Martin Hemberg</em></h4>
<h4 class="date"><em>2016-03-23</em></h4>
</div>
<!--bookdown:title:end-->

<!--bookdown:toc:start-->
  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">
<!--bookdown:toc2:start-->
<ul>
<li><a href="#about-the-course"><span class="toc-section-number">1</span> About the course</a><ul>
<li><a href="#registration"><span class="toc-section-number">1.1</span> Registration</a></li>
<li><a href="#github"><span class="toc-section-number">1.2</span> GitHub</a></li>
<li><a href="#license"><span class="toc-section-number">1.3</span> License</a></li>
<li><a href="#prerequisites"><span class="toc-section-number">1.4</span> Prerequisites</a></li>
</ul></li>
<li><a href="#technical-requirements"><span class="toc-section-number">2</span> Technical requirements</a></li>
<li><a href="#introduction-to-single-cell-rna-seq"><span class="toc-section-number">3</span> Introduction to single-cell RNA-seq</a><ul>
<li><a href="#bulk-rna-seq"><span class="toc-section-number">3.1</span> Bulk RNA-seq</a></li>
<li><a href="#scrna-seq"><span class="toc-section-number">3.2</span> scRNA-seq</a></li>
<li><a href="#protocol"><span class="toc-section-number">3.3</span> Protocol</a></li>
<li><a href="#analysis"><span class="toc-section-number">3.4</span> Analysis</a></li>
<li><a href="#challenges"><span class="toc-section-number">3.5</span> Challenges</a></li>
<li><a href="#controls"><span class="toc-section-number">3.6</span> Controls</a><ul>
<li><a href="#ercc-spike-ins"><span class="toc-section-number">3.6.1</span> ERCC <em>spike-ins</em></a></li>
<li><a href="#umis"><span class="toc-section-number">3.6.2</span> UMIs</a></li>
</ul></li>
</ul></li>
<li><a href="#construction-of-expression-matrix"><span class="toc-section-number">4</span> Construction of expression matrix</a><ul>
<li><a href="#reads-qc"><span class="toc-section-number">4.1</span> Reads QC</a></li>
<li><a href="#reads-alignment"><span class="toc-section-number">4.2</span> Reads alignment</a></li>
<li><a href="#alignment-example"><span class="toc-section-number">4.3</span> Alignment example</a></li>
<li><a href="#mapping-qc"><span class="toc-section-number">4.4</span> Mapping QC</a></li>
<li><a href="#reads-quantification"><span class="toc-section-number">4.5</span> Reads quantification</a></li>
</ul></li>
<li><a href="#scater-package"><span class="toc-section-number">5</span> scater package</a><ul>
<li><a href="#introduction"><span class="toc-section-number">5.1</span> Introduction</a></li>
<li><a href="#scater-workflow"><span class="toc-section-number">5.2</span> scater workflow</a></li>
<li><a href="#terminology"><span class="toc-section-number">5.3</span> Terminology</a></li>
<li><a href="#sceset-class"><span class="toc-section-number">5.4</span> SCESet class</a></li>
</ul></li>
<li><a href="#expression-qc-umi"><span class="toc-section-number">6</span> Expression QC (UMI)</a><ul>
<li><a href="#introduction-1"><span class="toc-section-number">6.1</span> Introduction</a></li>
<li><a href="#blischak-dataset"><span class="toc-section-number">6.2</span> Blischak dataset</a></li>
<li><a href="#gene-qc"><span class="toc-section-number">6.3</span> Gene QC</a><ul>
<li><a href="#gene-dropouts"><span class="toc-section-number">6.3.1</span> Gene dropouts</a></li>
<li><a href="#gene-expression"><span class="toc-section-number">6.3.2</span> Gene expression</a></li>
</ul></li>
<li><a href="#gene-filtering"><span class="toc-section-number">6.4</span> Gene filtering</a></li>
<li><a href="#cell-qc"><span class="toc-section-number">6.5</span> Cell QC</a><ul>
<li><a href="#library-size"><span class="toc-section-number">6.5.1</span> Library size</a></li>
<li><a href="#detected-genes-1"><span class="toc-section-number">6.5.2</span> Detected genes (1)</a></li>
<li><a href="#detected-genes-2"><span class="toc-section-number">6.5.3</span> Detected genes (2)</a></li>
<li><a href="#erccs-and-mts"><span class="toc-section-number">6.5.4</span> ERCCs and MTs</a></li>
</ul></li>
<li><a href="#cell-filtering"><span class="toc-section-number">6.6</span> Cell filtering</a><ul>
<li><a href="#default"><span class="toc-section-number">6.6.1</span> Default</a></li>
<li><a href="#automatic"><span class="toc-section-number">6.6.2</span> Automatic</a></li>
<li><a href="#manual"><span class="toc-section-number">6.6.3</span> Manual</a></li>
</ul></li>
<li><a href="#compare-filterings"><span class="toc-section-number">6.7</span> Compare filterings</a></li>
<li><a href="#save-the-data"><span class="toc-section-number">6.8</span> Save the data</a></li>
<li><a href="#exercise"><span class="toc-section-number">6.9</span> Exercise</a></li>
</ul></li>
<li><a href="#expression-qc-reads"><span class="toc-section-number">7</span> Expression QC (Reads)</a></li>
<li><a href="#expression-overview"><span class="toc-section-number">8</span> Expression overview</a><ul>
<li><a href="#introduction-2"><span class="toc-section-number">8.1</span> Introduction</a></li>
<li><a href="#top-500-genes"><span class="toc-section-number">8.2</span> Top 500 genes</a><ul>
<li><a href="#before-qc"><span class="toc-section-number">8.2.1</span> Before QC</a></li>
<li><a href="#after-qc"><span class="toc-section-number">8.2.2</span> After QC</a></li>
</ul></li>
<li><a href="#pca-plot"><span class="toc-section-number">8.3</span> PCA plot</a><ul>
<li><a href="#before-qc-1"><span class="toc-section-number">8.3.1</span> Before QC</a></li>
<li><a href="#after-qc-1"><span class="toc-section-number">8.3.2</span> After QC</a></li>
</ul></li>
<li><a href="#diffusion-map"><span class="toc-section-number">8.4</span> Diffusion map</a><ul>
<li><a href="#before-qc-2"><span class="toc-section-number">8.4.1</span> Before QC</a></li>
<li><a href="#after-qc-2"><span class="toc-section-number">8.4.2</span> After QC</a></li>
</ul></li>
<li><a href="#tsne-map"><span class="toc-section-number">8.5</span> tSNE map</a><ul>
<li><a href="#before-qc-3"><span class="toc-section-number">8.5.1</span> Before QC</a></li>
<li><a href="#after-qc-3"><span class="toc-section-number">8.5.2</span> After QC</a></li>
</ul></li>
<li><a href="#exercise-1"><span class="toc-section-number">8.6</span> Exercise</a></li>
</ul></li>
<li><a href="#expression-overview-reads"><span class="toc-section-number">9</span> Expression overview (Reads)</a></li>
<li><a href="#confounding-factors"><span class="toc-section-number">10</span> Confounding factors</a><ul>
<li><a href="#introduction-3"><span class="toc-section-number">10.1</span> Introduction</a></li>
<li><a href="#correlations-with-pcs"><span class="toc-section-number">10.2</span> Correlations with PCs</a><ul>
<li><a href="#detected-genes"><span class="toc-section-number">10.2.1</span> Detected genes</a></li>
<li><a href="#control-erccs"><span class="toc-section-number">10.2.2</span> Control ERCCs</a></li>
<li><a href="#control-mts"><span class="toc-section-number">10.2.3</span> Control MTs</a></li>
</ul></li>
<li><a href="#explanatory-variables"><span class="toc-section-number">10.3</span> Explanatory variables</a></li>
<li><a href="#other-confounders"><span class="toc-section-number">10.4</span> Other confounders</a></li>
<li><a href="#exercise-2"><span class="toc-section-number">10.5</span> Exercise</a></li>
</ul></li>
<li><a href="#confounding-factors-reads"><span class="toc-section-number">11</span> Confounding factors (Reads)</a></li>
<li><a href="#normalization-for-library-size"><span class="toc-section-number">12</span> Normalization for library size</a><ul>
<li><a href="#introduction-4"><span class="toc-section-number">12.1</span> Introduction</a></li>
<li><a href="#library-size-1"><span class="toc-section-number">12.2</span> Library size</a></li>
<li><a href="#normalisations"><span class="toc-section-number">12.3</span> Normalisations</a><ul>
<li><a href="#raw"><span class="toc-section-number">12.3.1</span> Raw</a></li>
<li><a href="#cpm"><span class="toc-section-number">12.3.2</span> CPM</a></li>
<li><a href="#log2cpm"><span class="toc-section-number">12.3.3</span> log2(CPM)</a></li>
<li><a href="#tmm"><span class="toc-section-number">12.3.4</span> TMM</a></li>
<li><a href="#rle"><span class="toc-section-number">12.3.5</span> RLE</a></li>
<li><a href="#upperquantile"><span class="toc-section-number">12.3.6</span> Upperquantile</a></li>
</ul></li>
<li><a href="#interpretation"><span class="toc-section-number">12.4</span> Interpretation</a></li>
<li><a href="#other-methods"><span class="toc-section-number">12.5</span> Other methods</a></li>
<li><a href="#visualize-genes"><span class="toc-section-number">12.6</span> Visualize genes</a><ul>
<li><a href="#raw-1"><span class="toc-section-number">12.6.1</span> Raw</a></li>
<li><a href="#cpm-1"><span class="toc-section-number">12.6.2</span> CPM</a></li>
<li><a href="#log2cpm-1"><span class="toc-section-number">12.6.3</span> log2(CPM)</a></li>
<li><a href="#upperquantile-1"><span class="toc-section-number">12.6.4</span> Upperquantile</a></li>
<li><a href="#fpkm"><span class="toc-section-number">12.6.5</span> FPKM</a></li>
<li><a href="#tpm"><span class="toc-section-number">12.6.6</span> TPM</a></li>
</ul></li>
<li><a href="#exercise-3"><span class="toc-section-number">12.7</span> Exercise</a></li>
</ul></li>
<li><a href="#normalization-for-library-size-reads"><span class="toc-section-number">13</span> Normalization for library size (Reads)</a></li>
<li><a href="#remove-confounders-using-controls"><span class="toc-section-number">14</span> Remove confounders using controls</a><ul>
<li><a href="#introduction-5"><span class="toc-section-number">14.1</span> Introduction</a></li>
<li><a href="#brennecke-method"><span class="toc-section-number">14.2</span> Brennecke method</a></li>
<li><a href="#remove-unwanted-variation"><span class="toc-section-number">14.3</span> Remove Unwanted Variation</a></li>
<li><a href="#effectiveness-1"><span class="toc-section-number">14.4</span> Effectiveness 1</a></li>
<li><a href="#effectiveness-2"><span class="toc-section-number">14.5</span> Effectiveness 2</a></li>
<li><a href="#exercise-4"><span class="toc-section-number">14.6</span> Exercise</a></li>
</ul></li>
<li><a href="#remove-confounders-using-controls-reads"><span class="toc-section-number">15</span> Remove confounders using controls (Reads)</a></li>
</ul>
<!--bookdown:toc2:end-->
      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Analysis of single-cell RNA-seq data</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:toc:end-->
<!--bookdown:body:start-->
<div id="about-the-course" class="section level1">
<h1><span class="header-section-number">1</span> About the course</h1>
<p>Recent technological advances have made it possible to obtain genome-wide transcriptome data from single cells using high-throughput sequencing (scRNA-seq). Even though scRNA-seq makes it possible to address problems that are intractable with bulk RNA-seq data, analysing scRNA-seq is also more challenging.</p>
<p>In this course we will be surveying the existing problems as well as the available computational and statistical frameworks available for the analysis of scRNA-seq.</p>
<div id="registration" class="section level2">
<h2><span class="header-section-number">1.1</span> Registration</h2>
<p><a href="http://training.csx.cam.ac.uk/bioinformatics/event/1626698" target="blank">http://training.csx.cam.ac.uk/bioinformatics/event/1626698</a></p>
</div>
<div id="github" class="section level2">
<h2><span class="header-section-number">1.2</span> GitHub</h2>
<p><a href="https://github.com/hemberg-lab/scRNA.seq.course" target="blank">https://github.com/hemberg-lab/scRNA.seq.course</a></p>
</div>
<div id="license" class="section level2">
<h2><span class="header-section-number">1.3</span> License</h2>
<p><a href="http://creativecommons.org/licenses/by-nd/4.0" target="blank">Creative Commons Attribution-NoDerivatives 4.0 International License</a></p>
</div>
<div id="prerequisites" class="section level2">
<h2><span class="header-section-number">1.4</span> Prerequisites</h2>
<p>The course is intended for those who have basic familiarity with Unix and the R scripting language.</p>
<p>We will also assume that you are familiar with mapping and analysing bulk RNA-seq data as well as with the commonly available computational tools.</p>
<p>We recommend attending the <a href="http://training.csx.cam.ac.uk/bioinformatics/event/1544785">Introduction to RNA-seq and ChIP-seq data analysis</a> or the <a href="http://training.csx.cam.ac.uk/bioinformatics/event/1614561">Analysis of high-throughput sequencing data with Bioconductor</a> before attending this course.</p>
<!--chapter:end:index.Rmd-->
</div>
</div>
<div id="technical-requirements" class="section level1">
<h1><span class="header-section-number">2</span> Technical requirements</h1>
<p>We tried to make this course purely <a href="https://www.r-project.org/">R-based</a>. However, one of the methods that we describe (<a href="http://bioinfo.uncc.edu/SNNCliq/">SNN-Cliq</a>) is only partly R-based. It makes a simple <em>python</em> call from R and requires a user to have rights to write to the current directory. You also need to download <a href="http://bioinfo.uncc.edu/SNNCliq/Cliq.txt">this file</a> and put it in <code>~/snn-cliq/</code> directory.</p>
<p>Other than that, before running the course exercises, all you need to do is to install the following packages in your R:</p>
<p><a href="https://cran.r-project.org/web/packages/devtools/index.html">devtools</a> for installing packages from GitHub:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;devtools&quot;</span>)</code></pre></div>
<p><a href="https://github.com/hemberg-lab/scRNA.seq.funcs">scRNA.seq.funcs</a> - R package containing our own functions used in this course:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&quot;hemberg-lab/scRNA.seq.funcs&quot;</span>)</code></pre></div>
<p><a href="https://cran.r-project.org/web/packages/mvoutlier/index.html">mvoutlier</a> - for an automatic outlier detection in the <a href="https://github.com/davismcc/scater">scater</a> package.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;mvoutlier&quot;</span>)</code></pre></div>
<p><a href="https://github.com/tallulandrews/M3D">M3D</a> for identification of important and DE genes, developed by <a href="http://www.sanger.ac.uk/people/directory/andrews-tallulah-s">Tallulah Andrews</a>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&quot;tallulandrews/M3D&quot;</span>, <span class="dt">ref =</span> <span class="st">&quot;release&quot;</span>)</code></pre></div>
<p><a href="https://bioconductor.org/packages/release/bioc/html/RUVSeq.html">RUVSeq</a> for normalization using ERCC controls:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;RUVSeq&quot;</span>)</code></pre></div>
<p><a href="https://cran.r-project.org/web/packages/ROCR/index.html">ROCR</a> for performance estimations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;ROCR&quot;</span>)</code></pre></div>
<p><a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> for plotting Venn diagrams:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;limma&quot;</span>)</code></pre></div>
<p><a href="https://bioconductor.org/packages/release/bioc/html/DESeq2.html">DESeq2</a> for calculation of DE genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;DESeq2&quot;</span>)</code></pre></div>
<p><a href="http://hms-dbmi.github.io/scde/">scde</a> for calculation of DE genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&quot;hms-dbmi/scde&quot;</span>, <span class="dt">build_vignettes =</span> <span class="ot">FALSE</span>)</code></pre></div>
<ul>
<li><p>Installation on Mac OS X may require <a href="http://thecoatlessprofessor.com/programming/rcpp-rcpparmadillo-and-os-x-mavericks-lgfortran-and-lquadmath-error/">this additional gfortran library</a>:</p>
<pre><code>curl -O http://r.research.att.com/libs/gfortran-4.8.2-darwin13.tar.bz2
sudo tar fvxz gfortran-4.8.2-darwin13.tar.bz2 -C /</code></pre></li>
<li><p>See the <a href="http://hms-dbmi.github.io/scde/help.html">help page</a> for additional support.</p></li>
</ul>
<p><a href="https://cran.r-project.org/web/packages/pheatmap/index.html">pheatmap</a> for plotting good looking heatmaps:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;pheatmap&quot;</span>)</code></pre></div>
<p><a href="http://bioconductor.org/packages/release/bioc/html/pcaMethods.html">pcaMethods</a> required by <strong>pcaReduce</strong> package below for unsupervised clustering of scRNA-seq data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">## try http:// if https:// URLs are not supported
<span class="kw">source</span>(<span class="st">&quot;https://bioconductor.org/biocLite.R&quot;</span>)
<span class="kw">biocLite</span>(<span class="st">&quot;pcaMethods&quot;</span>)</code></pre></div>
<p><a href="https://github.com/JustinaZ/pcaReduce">pcaReduce</a> for unsupervised clustering of scRNA-seq data (<a href="http://biorxiv.org/content/early/2015/09/08/026385">bioRxiv</a>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&quot;JustinaZ/pcaReduce&quot;</span>)</code></pre></div>
<p><a href="https://cran.r-project.org/web/packages/pheatmap/index.html">Rtsne</a> for unsupervised clustering of scRNA-seq data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">install.packages</span>(<span class="st">&quot;Rtsne&quot;</span>)</code></pre></div>
<p><a href="http://bioconductor.org/packages/devel/bioc/html/SC3.html">SC3</a> for unsupervised clustering of scRNA-seq data (<a href="http://biorxiv.org/content/early/2016/01/13/036558">bioRxiv</a>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">devtools::<span class="kw">install_github</span>(<span class="st">&quot;hemberg-lab/SC3&quot;</span>, <span class="dt">ref =</span> <span class="st">&quot;R-old&quot;</span>)</code></pre></div>
<ul>
<li>Before running <strong>SC3</strong> for the first time <strong>only</strong>, please start R and enter:</li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">RSelenium::<span class="kw">checkForServer</span>()</code></pre></div>
<!--chapter:end:01-reqs.Rmd-->
</div>
<div id="introduction-to-single-cell-rna-seq" class="section level1">
<h1><span class="header-section-number">3</span> Introduction to single-cell RNA-seq</h1>
<div id="bulk-rna-seq" class="section level2">
<h2><span class="header-section-number">3.1</span> Bulk RNA-seq</h2>
<ul>
<li>A major breakthrough (replaced microarrays) in the late 00’s and has been widely used since</li>
<li>Provides an <strong>average expression level</strong> for each gene from a large population of input cells</li>
<li>Useful for comparative transcriptomics, e.g. samples of the same tissue from different species</li>
<li>Useful for quantifying expression signatures from ensembles, e.g. in disease studies</li>
<li><strong>Insufficient</strong> for studying heterogeneous systems, e.g. early development studies, complex tissues (brain)</li>
<li>Does <strong>not</strong> provide insights into the stochastic nature of gene expression</li>
</ul>
</div>
<div id="scrna-seq" class="section level2">
<h2><span class="header-section-number">3.2</span> scRNA-seq</h2>
<ul>
<li>A <strong>new</strong> technology, first publication by <a href="http://www.nature.com/nmeth/journal/v6/n5/abs/nmeth.1315.html">Tang et al</a> in 2009</li>
<li>Instead of providing an average of expression of a population of cells, scRNA-seq provides a <strong>distribution of expression levels</strong> from a population of cells</li>
<li>Allows to study new biological questions in which <strong>cell-specific changes in transcriptome are important</strong>, e.g. cell type identification, inference of gene regulatory networks across the cells, stochastic component of transcription</li>
<li>Datasets range <strong>from <span class="math inline">\(10^2\)</span> to <span class="math inline">\(10^4\)</span> cells</strong> and increase in size every year</li>
<li>Currently there are several different protocols in use, e.g. <a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2639.html">SMART-seq2</a>, <a href="http://www.cell.com/cell-reports/abstract/S2211-1247%2812%2900228-8">CELL-seq</a> and <a href="http://mccarrolllab.com/dropseq/">Drop-seq</a></li>
<li>Several computational analysis methods from bulk RNA-seq <strong>can</strong> be used</li>
<li><strong>In most cases</strong> requires adaptation of the existing methods or development of new ones</li>
</ul>
</div>
<div id="protocol" class="section level2">
<h2><span class="header-section-number">3.3</span> Protocol</h2>
<div class="figure">
<img src="figures/RNA-Seq_workflow-5.pdf.jpg" />

</div>
<p>image from <a href="https://en.wikipedia.org/wiki/Single_cell_sequencing">Wikipedia - Single cell sequencing</a></p>
<p>Overall, experimental scRNA-seq protocols are similar to the methods used for bulk RNA-seq. For a discussion on experimental methods, please see reviews by <a href="http://nar.oxfordjournals.org/content/42/14/8845">Saliba et al</a>, <a href="http://www.sciencedirect.com/science/article/pii/S1097276515003068">Handley et al</a> or <a href="http://www.sciencedirect.com/science/article/pii/S1097276515002610">Kolodziejczyk et al</a>.</p>
</div>
<div id="analysis" class="section level2">
<h2><span class="header-section-number">3.4</span> Analysis</h2>
<p>This course is concerned with the computational analysis of the data obtained from scRNA-seq experiments. Even though the format of the data is the same as for bulk RNA-seq, the fact that individual cells are sampled means that the data <strong>should</strong> be analyzed differently. We will provide an overview of how to process a scRNA-seq sample and how to analyze the data to provide biological insights.</p>
<div class="figure">
<img src="figures/flowchart.png" />

</div>
<p>Flowchart for analyzing scRNA-seq data.</p>
</div>
<div id="challenges" class="section level2">
<h2><span class="header-section-number">3.5</span> Challenges</h2>
<p>The main difference between bulk and single cell RNA-seq is that each sequencing library represents a single cell, instead of a population of cells. Therefore, significant attention has to be paid to comparison of the results from different cells (sequencing libraries). The main sources of discrepancy between the libraries are:</p>
<ul>
<li><strong>Amplification</strong> (up to 1 million fold)</li>
<li><strong>Gene ‘dropouts’</strong> in which a gene is observed at a moderate expression level in one cell but is not detected in another cell (<a href="http://www.nature.com/nmeth/journal/v11/n7/full/nmeth.2967.html">Kharchenko et al</a>).</li>
</ul>
<p>In both cases the discrepancies are introduced due to low starting transcript amount (from one cell only) and have yet to be addressed by improving the experimental protocol.</p>
</div>
<div id="controls" class="section level2">
<h2><span class="header-section-number">3.6</span> Controls</h2>
<p>To address challenges in technical variation between scRNA sequencing libraries two quantitative standards were introduced. They allow to normalize gene expression levels across different cells.</p>
<div id="ercc-spike-ins" class="section level3">
<h3><span class="header-section-number">3.6.1</span> ERCC <em>spike-ins</em></h3>
<p>One standard, strongly recommended for all scRNA-seq experiments is to use extrinsic <em>spike-in</em> molecules. These molecules are added to the lysate before the reverse transcription. The most popular and widely used artificial spike-ins are from <a href="https://www.thermofisher.com/order/catalog/product/4456740">External RNA Control Consortium (ERCC)</a>. It contains 92 synthetic spikes based on bacterial sequences. Normalization using <em>spike-ins</em> is based on the fact that the number of molecules of each <em>spike-in</em> RNA species should be the same across all single-cell libraries.</p>
</div>
<div id="umis" class="section level3">
<h3><span class="header-section-number">3.6.2</span> UMIs</h3>
<p>Another method of standartisation is to use <a href="http://www.nature.com/nmeth/journal/v9/n1/full/nmeth.1778.html">Unique Molecular Identifiers (UMIs)</a>. Instead of sequencing the amplified reads from a cell library, it allows for sequencing reads derived solely from 3’ or 5’ end of the amplified transcript. UMIs are added as barcodes to the individual RNA molecules. This approach provides an estimate of the number of transcripts that is independent of amplification biases.</p>
<!--chapter:end:02-intro.Rmd-->
</div>
</div>
</div>
<div id="construction-of-expression-matrix" class="section level1">
<h1><span class="header-section-number">4</span> Construction of expression matrix</h1>
<div id="reads-qc" class="section level2">
<h2><span class="header-section-number">4.1</span> Reads QC</h2>
<p>The output from a scRNA-seq experiment is a large collection of short cDNA reads. The first step is to ensure that the reads are of high quality. In application to scRNA-seq, this can be performed by using standard bulk RNA-seq QC tools, such as <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a> or <a href="http://www.ebi.ac.uk/research/enright/software/kraken">Kraken</a>. Currently, there are no specialized tools for scRNA-seq available, so we use <a href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">FastQC</a>.</p>
<p>Assuming that our reads are in experiment.bam, we run FastQC as</p>
<pre><code>$&lt;path_to_fastQC&gt;/fastQC experiment.bam</code></pre>
<p>Below is an example of the output from FastQC for a dataset of 125 bp reads. Here, the quality of the reads is overall high, so we can proceed the analysis with confidence.</p>
<div class="figure">
<img src="figures/per_base_quality.png" />

</div>
<p>Additionally, the data can be visualized using the <a href="https://www.broadinstitute.org/igv/">Integrative Genomics Browser (IGV)</a>.</p>
</div>
<div id="reads-alignment" class="section level2">
<h2><span class="header-section-number">4.2</span> Reads alignment</h2>
<p>Once the low-quality reads have been removed, the remaining reads can be mapped to a reference genome. Again, there are no special purpose methods for this, so we can use the <a href="https://github.com/alexdobin/STAR">STAR</a> or the <a href="https://ccb.jhu.edu/software/tophat/index.shtml">TopHat</a> aligner.</p>
<p>An example of how to map reads.bam to using STAR hg38 is</p>
<pre><code>$&lt;path_to_STAR&gt;/STAR --runThreadN 1 --runMode alignReads
--readFilesIn reads1.fq.gz reads2.fq.gz --readFilesCommand zcat --genomeDir &lt;path&gt;
--parametersFiles FileOfMoreParameters.txt --outFileNamePrefix &lt;outpath&gt;/output</code></pre>
<p><strong>Note</strong>, if the <em>spike-ins</em> are used, the reference sequence should be augmented with the DNA sequence of the <em>spike-in</em> molecules before mapping.</p>
<p><strong>Note</strong>, when UMIs are used, their barcodes should be removed from every read.</p>
<p>Once we have mapped the reads for each cell to the reference genome, we need to make sure that a sufficient number of reads from each cell could be mapped to the reference genome. In our experience, the fraction of mappable reads is 60-70%. However, this result may vary depending on protocol, read length and settings for the read alignment. As a general rule, we expect all cells to have a similar fraction of mapped reads, so any outliers should be inspected and possibly removed.</p>
</div>
<div id="alignment-example" class="section level2">
<h2><span class="header-section-number">4.3</span> Alignment example</h2>
<p>The histogram below shows the total number of reads mapped to each cell for an scRNA-seq experiment. Each bar represents one cell, and they have been sorted in ascending order by the total number of reads per cell. The three red arrows indicate cells that are outliers in terms of their coverage and they should be removed from further analysis. The two yellow arrows point to cells with a surprisingly large number of unmapped reads. However, we deem the discrepancy as small and we retain the cells for now.</p>
<div class="figure">
<img src="figures/Bergiers_exp1_mapping_by_cell.png" />

</div>
</div>
<div id="mapping-qc" class="section level2">
<h2><span class="header-section-number">4.4</span> Mapping QC</h2>
<p>After mapping the raw sequencing to the genome we should evaluate the quality of the mapping. There are many ways to measure this including: amount of reads mapping to rRNA/tRNAs, proportion of uniquely mapping reads, reads mapping across splice junctions, read depth along the transcripts. Methods developed for bulk RNA-seq, such as <a href="http://rseqc.sourceforge.net/">RSeQC</a>, are applicable to single-cell data:</p>
<pre><code>python &lt;RSeQCpath&gt;/geneBody_coverage.py -i input.bam -r genome.bed -o output.txt
python &lt;RSeQCpath&gt;/bam_stat.py -i input.bam -r genome.bed -o output.txt
python &lt;RSeQCpath&gt;/split_bam.py -i input.bam -r rRNAmask.bed -o output.txt</code></pre>
<p>However the expected results will depend on the experimental protocol, e.g. many scRNA-seq methods use poly-A selection to avoid sequencing rRNAs which results in a 3’ bias in the read coverage across the genes (aka gene body coverage). The figure below shows this 3’ bias as well as three cells which were outliers and removed from the dataset:</p>
<div class="figure">
<img src="figures/Exp1_RSEQC_geneBodyCoverage_plot_Combined.png" />

</div>
</div>
<div id="reads-quantification" class="section level2">
<h2><span class="header-section-number">4.5</span> Reads quantification</h2>
<p>The next step is to quantify the expression level of each gene for each cell. For mRNA data, we can use one of the tools which has been developed for bulk RNA-seq data, e.g. <a href="http://www-huber.embl.de/users/anders/HTSeq/">HT-seq</a> or <a href="http://subread.sourceforge.net/">FeatureCounts</a></p>
<pre><code># include multimapping
&lt;featureCounts_path&gt;/featureCounts -O -M -Q 30 -p -a genome.gtf -o outputfile input.bam
# exclude multimapping
&lt;featureCounts_path&gt;/featureCounts -Q 30 -p -a genome.gtf -o outputfile input.bam</code></pre>
<p><strong>Note</strong>, when UMIs are used, the expression counts can be collapsed by summing the number of unique barcodes associated with all reads mapped to a given gene.</p>
<!--chapter:end:03-exprs-constr.Rmd-->
</div>
</div>
<div id="scater-package" class="section level1">
<h1><span class="header-section-number">5</span> scater package</h1>
<div id="introduction" class="section level2">
<h2><span class="header-section-number">5.1</span> Introduction</h2>
<p><a href="https://github.com/davismcc/scater">scater</a> is a single-cell analysis toolkit for expression with R. This package contains useful tools for the analysis of single-cell gene expression data using the statistical software R. The package places an emphasis on tools for quality control, visualisation and pre-processing of data before further downstream analysis.</p>
<p><a href="https://github.com/davismcc/scater">scater</a> enables the following:</p>
<ul>
<li>Automated computation of QC metrics</li>
<li>Transcript quantification from read data with pseudo-alignment</li>
<li>Data format standardisation</li>
<li>Rich visualisations for exploratory analysis</li>
<li>Seamless integration into the Bioconductor universe</li>
<li>Simple normalisation methods</li>
</ul>
<p>We highly recommend to use <a href="https://github.com/davismcc/scater">scater</a> for any single-cell RNA-seq analysis you will be doing in the future. <a href="https://github.com/davismcc/scater">scater</a> is the basis of the first part of the course and therefore we will spend some time explaining its details.</p>
</div>
<div id="scater-workflow" class="section level2">
<h2><span class="header-section-number">5.2</span> scater workflow</h2>
<div class="figure">
<img src="figures/scater_qc_workflow.png" />

</div>
</div>
<div id="terminology" class="section level2">
<h2><span class="header-section-number">5.3</span> Terminology</h2>
<p>(this chapter is taken from the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/scater/inst/doc/vignette.html">scater vignette</a>)</p>
<ul>
<li>The capabilities of scater are built on top of Bioconductor’s Biobase.</li>
<li>In Bioconductor terminology we assay numerous <strong>“features”</strong> for a number of <strong>“samples”</strong>.</li>
<li>Features, in the context of <a href="https://github.com/davismcc/scater">scater</a>, correspond most commonly to genes or transcripts, but could be any general genomic or transcriptomic regions (e.g. exon) of interest for which we take measurements.</li>
<li>In the following chapters it may be more intuitive to mentally replace <strong>“feature”</strong> with <strong>“gene”</strong> or <strong>“transcript”</strong> (depending on the context of the study) wherever <strong>“feature”</strong> appears.</li>
<li>In the scater context, <strong>“samples”</strong> refer to individual cells that we have assayed.</li>
</ul>
</div>
<div id="sceset-class" class="section level2">
<h2><span class="header-section-number">5.4</span> SCESet class</h2>
<p>(this chapter is taken from the <a href="https://www.bioconductor.org/packages/devel/bioc/vignettes/scater/inst/doc/vignette.html">scater vignette</a>)</p>
<p>In scater we organise single-cell expression data in objects of the <strong>SCESet</strong> class. The class inherits the Bioconductor ExpressionSet class, which provides a common interface familiar to those who have analyzed microarray experiments with Bioconductor. The class requires three input files:</p>
<ul>
<li><strong>exprs</strong>, a numeric matrix of expression values, where rows are features, and columns are cells</li>
<li><strong>phenoData</strong>, an <em>AnnotatedDataFrame</em> object, where rows are cells, and columns are cell attributes (such as cell type, culture condition, day captured, etc.)</li>
<li><strong>featureData</strong>, an <em>AnnotatedDataFrame</em> object, where rows are features (e.g. genes), and columns are feature attributes, such as biotype, gc content, etc.</li>
</ul>
<p>For more details about other features inherited from Bioconductor’s <strong>ExpressionSet</strong> class, type <code>?ExpressionSet</code> at the R prompt.</p>
<p>When your data is wrapped in the <strong>SCESet</strong> class, <a href="https://github.com/davismcc/scater">scater</a> will do the dirty job of calculating different properties of the data automatically. You will see how it works in the next chapters.</p>
<!--chapter:end:04-scater.Rmd-->
</div>
</div>
<div id="expression-qc-umi" class="section level1">
<h1><span class="header-section-number">6</span> Expression QC (UMI)</h1>
<div id="introduction-1" class="section level2">
<h2><span class="header-section-number">6.1</span> Introduction</h2>
<p>Once gene expression has been quantified the resulting expression matrix must be examined for poor quality cells which were not detected from QC of the raw reads. Failure to remove low quality cells at this stage will add a large amount of technical noise which will obscure the biological signals of interest in downstream analysis. We will demonstrate some simple filters that can be applied directly to the gene expression matrix with minimal extrinsic information.</p>
<p>Typically, the results are summarized using an <strong>expression matrix</strong>. In the expression matrix, each row represents a gene and each column represents a cell.</p>
<p>To illustrate cell QC, we consider a <a href="http://jdblischak.github.io/singleCellSeq/analysis/">dataset</a> from lymphoblast and induced pluripotent stem cells generated by John Blischak in <a href="http://giladlab.uchicago.edu/">Yoav Gilad</a>’s lab at the University of Chicago. The experiments were carried out on the Fluidigm C1 platform and to facilitate the quantification both unique molecular identifiers (UMIs) and ERCC <em>spike-ins</em> were used. For our purposes you need to download the files <code>annotation.txt</code>, <code>molecules.txt</code>, and <code>reads.txt</code> from <a href="https://github.com/hemberg-lab/scRNA.seq.course/tree/master/inst/materials/blischak">here</a> into the <code>blischak</code> folder in your working directory. These files are the copies of the original files made on the 15/03/16. We will use these copies for reproducibility purposes.</p>
<p>To illustrate how one can standardise the analysis of scRNA-seq data we will also be using the <a href="https://github.com/davismcc/scater">scater</a> package, described in the previous chapter.</p>
</div>
<div id="blischak-dataset" class="section level2">
<h2><span class="header-section-number">6.2</span> Blischak dataset</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(scater, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)
<span class="kw">library</span>(knitr)
<span class="kw">options</span>(<span class="dt">stringsAsFactors =</span> <span class="ot">FALSE</span>)
opts_chunk$<span class="kw">set</span>(<span class="dt">out.width=</span><span class="st">&#39;90%&#39;</span>, <span class="dt">fig.align =</span> <span class="st">&#39;center&#39;</span>)</code></pre></div>
<p>Load the data and annotations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">molecules &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;blischak/molecules.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)
anno &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;blischak/annotation.txt&quot;</span>, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>, <span class="dt">header =</span> <span class="ot">TRUE</span>)</code></pre></div>
<p>How does the data look like?</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
    <span class="kw">head</span>(molecules[ , <span class="dv">1</span>:<span class="dv">3</span>]), <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
    <span class="dt">caption =</span> <span class="st">&#39;A table of the first 6 rows and 3 columns of the molecules table.&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-18)A table of the first 6 rows and 3 columns of the molecules table.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">NA19098.r1.A01</th>
<th align="right">NA19098.r1.A02</th>
<th align="right">NA19098.r1.A03</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ENSG00000237683</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>ENSG00000187634</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>ENSG00000188976</td>
<td align="right">3</td>
<td align="right">6</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>ENSG00000187961</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>ENSG00000187583</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>ENSG00000187642</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
    <span class="kw">head</span>(anno), <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
    <span class="dt">caption =</span> <span class="st">&#39;A table of the first 6 rows of the anno table.&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-18)A table of the first 6 rows of the anno table.</caption>
<thead>
<tr class="header">
<th align="left">individual</th>
<th align="left">replicate</th>
<th align="left">well</th>
<th align="left">batch</th>
<th align="left">sample_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A01</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A01</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A02</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A02</td>
</tr>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A03</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A03</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A04</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A04</td>
</tr>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A05</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A05</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A06</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A06</td>
</tr>
</tbody>
</table>
<p>The data consists of 3 individuals and 3 replicates and therefore has 9 batches in total.</p>
<p>Let’s standardise the analysis by using the scater package described above. First, make scater SCESet classes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">pheno_data &lt;-<span class="st"> </span><span class="kw">new</span>(<span class="st">&quot;AnnotatedDataFrame&quot;</span>, anno)
<span class="kw">rownames</span>(pheno_data) &lt;-<span class="st"> </span>pheno_data$sample_id
umi &lt;-<span class="st"> </span>scater::<span class="kw">newSCESet</span>(
    <span class="dt">countData =</span> molecules,
    <span class="dt">phenoData =</span> pheno_data
)</code></pre></div>
<p>Remove genes that are not expressed in any cell:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">keep_feature &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">is_exprs</span>(umi)) &gt;<span class="st"> </span><span class="dv">0</span>
umi &lt;-<span class="st"> </span>umi[keep_feature, ]</code></pre></div>
<p>Define control features (genes) - ERCC spike-ins and mitochondrial genes (<a href="http://jdblischak.github.io/singleCellSeq/analysis/qc-filter-ipsc.html">provided</a> by the authors):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">ercc &lt;-<span class="st"> </span><span class="kw">featureNames</span>(umi)[<span class="kw">grepl</span>(<span class="st">&quot;ERCC-&quot;</span>, <span class="kw">featureNames</span>(umi))]
mt &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;ENSG00000198899&quot;</span>, <span class="st">&quot;ENSG00000198727&quot;</span>, <span class="st">&quot;ENSG00000198888&quot;</span>,
        <span class="st">&quot;ENSG00000198886&quot;</span>, <span class="st">&quot;ENSG00000212907&quot;</span>, <span class="st">&quot;ENSG00000198786&quot;</span>,
        <span class="st">&quot;ENSG00000198695&quot;</span>, <span class="st">&quot;ENSG00000198712&quot;</span>, <span class="st">&quot;ENSG00000198804&quot;</span>,
        <span class="st">&quot;ENSG00000198763&quot;</span>, <span class="st">&quot;ENSG00000228253&quot;</span>, <span class="st">&quot;ENSG00000198938&quot;</span>,
        <span class="st">&quot;ENSG00000198840&quot;</span>)</code></pre></div>
<p>Calculate the quality metrics:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi &lt;-<span class="st"> </span>scater::<span class="kw">calculateQCMetrics</span>(
    umi,
    <span class="dt">feature_controls =</span> <span class="kw">list</span>(<span class="dt">ERCC =</span> ercc, <span class="dt">MT =</span> mt)
)</code></pre></div>
</div>
<div id="gene-qc" class="section level2">
<h2><span class="header-section-number">6.3</span> Gene QC</h2>
<div id="gene-dropouts" class="section level3">
<h3><span class="header-section-number">6.3.1</span> Gene dropouts</h3>
<p>First, one can look at the gene expression frequency versus mean expression level to assess the effects of technical dropout in the dataset.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scater::<span class="kw">plotQC</span>(umi, <span class="dt">type =</span> <span class="st">&quot;exprs-freq-vs-mean&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/dropout-overview-1.png" alt="(\#fig:dropout-overview)Dropout rate vs mean expression" width="90%" />
<p class="caption">
(#fig:dropout-overview)Dropout rate vs mean expression
</p>
</div>
<p>Interestingly, only less than half of the genes are expressed in more than 50% of the cells.</p>
</div>
<div id="gene-expression" class="section level3">
<h3><span class="header-section-number">6.3.2</span> Gene expression</h3>
<p>One can also look at the number of reads consumed by the top 50 expressed genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scater::<span class="kw">plotQC</span>(umi, <span class="dt">type =</span> <span class="st">&quot;highest-expression&quot;</span>)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/top50-gene-expr-1.png" alt="(\#fig:top50-gene-expr)Number of total counts consumed by the top 50 expressed genes" width="90%" />
<p class="caption">
(#fig:top50-gene-expr)Number of total counts consumed by the top 50 expressed genes
</p>
</div>
<p>Not surprisingly, most of the very top expressed genes are either ERCCs or MT genes (Feature controls).</p>
</div>
</div>
<div id="gene-filtering" class="section level2">
<h2><span class="header-section-number">6.4</span> Gene filtering</h2>
<p>We can remove genes whose expression level is considered <strong>“undetectable”</strong>. We define a detectable gene expression level, in which at least 2 cells contain more than 1 read mapping to the gene.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_genes &lt;-<span class="st"> </span><span class="kw">apply</span>(<span class="kw">counts</span>(umi), <span class="dv">1</span>, function(x) <span class="kw">length</span>(x[x &gt;<span class="st"> </span><span class="dv">1</span>]) &gt;=<span class="st"> </span><span class="dv">2</span>)
<span class="kw">fData</span>(umi)$use &lt;-<span class="st"> </span>filter_genes</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
    <span class="kw">as.data.frame</span>(<span class="kw">table</span>(filter_genes)),
    <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
    <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
    <span class="dt">caption =</span> <span class="st">&#39;The number of genes removed by gene filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-24)The number of genes removed by gene filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_genes</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">4512</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">14214</td>
</tr>
</tbody>
</table>
<p>Depending on the cell-type, protocol and sequencing depth, other cut-offs may be appropriate.</p>
</div>
<div id="cell-qc" class="section level2">
<h2><span class="header-section-number">6.5</span> Cell QC</h2>
<div id="library-size" class="section level3">
<h3><span class="header-section-number">6.5.1</span> Library size</h3>
<p>Next we will look at the total number of RNA molecules detected per sample (if we were using read counts rather than UMI counts this would be total reads). Wells with few reads/molecules are likely to have been broken or failed to capture a cell, thus should be removed.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    umi$total_counts,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-counts-hist-1.png" alt="(\#fig:total-counts-hist)Histogram of library sizes for all cells" width="90%" />
<p class="caption">
(#fig:total-counts-hist)Histogram of library sizes for all cells
</p>
</div>
<p><strong>Exercise</strong></p>
<p>Apply a suitable filter to remove the cells that contain too few molecules. What distribution do you expect that the total number of molecules for each cell should follow?</p>
<p><strong>Answer</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_total_counts &lt;-<span class="st"> </span>(umi$total_counts &gt;<span class="st"> </span><span class="dv">10000</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
    <span class="kw">as.data.frame</span>(<span class="kw">table</span>(filter_by_total_counts)),
    <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
    <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
    <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by total counts filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-26)The number of cells removed by total counts filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_total_counts</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">862</td>
</tr>
</tbody>
</table>
</div>
<div id="detected-genes-1" class="section level3">
<h3><span class="header-section-number">6.5.2</span> Detected genes (1)</h3>
<p>One of the simplest measures of cell quality is the number of genes that were detected. Cells with few detected genes may have been broken or dead prior to capture.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">hist</span>(
    umi$total_features,
    <span class="dt">breaks =</span> <span class="dv">100</span>
)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-features-hist-1.png" alt="(\#fig:total-features-hist)Histogram of the number of detected genes in all cells" width="90%" />
<p class="caption">
(#fig:total-features-hist)Histogram of the number of detected genes in all cells
</p>
</div>
<p>Here we see that most cells have between 5,000-11,000 detected genes, which is normal for high-depth scRNA-seq. However this varies by experimental protocol and sequencing depth with droplet-based methods or lower sequencing-depth detecting fewer genes per cell. The feature to note is the <strong>“heavy tail”</strong> of the left hand side of the distribution. If detection rates were equal across the cells then the distribution should be approximately normal. Thus we remove those cells in the tail of the distribution (&lt;5,000 detected genes).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_expr_features &lt;-<span class="st"> </span>(umi$total_features &gt;<span class="st"> </span><span class="dv">5000</span>)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
    <span class="kw">as.data.frame</span>(<span class="kw">table</span>(filter_by_expr_features)),
    <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
    <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
    <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by total features filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-28)The number of cells removed by total features filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_expr_features</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">16</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">848</td>
</tr>
</tbody>
</table>
</div>
<div id="detected-genes-2" class="section level3">
<h3><span class="header-section-number">6.5.3</span> Detected genes (2)</h3>
<p>Additionally, we can plot the library size as a function of the number of the detected genes. This will also allow us to find other cell outliers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scater::<span class="kw">plotPhenoData</span>(
    umi,
    <span class="kw">aes</span>(<span class="dt">x =</span> total_features, <span class="dt">y =</span> <span class="kw">log10</span>(total_counts), <span class="dt">colour =</span> batch)
)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-features-vs-counts-1.png" alt="(\#fig:total-features-vs-counts)Library size vs number of detected genes" width="90%" />
<p class="caption">
(#fig:total-features-vs-counts)Library size vs number of detected genes
</p>
</div>
</div>
<div id="erccs-and-mts" class="section level3">
<h3><span class="header-section-number">6.5.4</span> ERCCs and MTs</h3>
<p>Another measures of cell quality is the ration between ERCC <em>spike-in</em> RNAs and endogenous RNAs. This can be used to estimate the total amount of RNA in the captured cells. Cells with a high level of <em>spike-in</em> RNAs had low starting amounts of RNA, likely due to the cell being dead or stressed, or the RNA being degraded.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scater::<span class="kw">plotPhenoData</span>(
    umi,
    <span class="kw">aes_string</span>(<span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
               <span class="dt">y =</span> <span class="st">&quot;pct_counts_feature_controls_MT&quot;</span>,
               <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>)
)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/mt-vs-counts-1.png" alt="(\#fig:mt-vs-counts)Percentage of counts in MT genes" width="90%" />
<p class="caption">
(#fig:mt-vs-counts)Percentage of counts in MT genes
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">scater::<span class="kw">plotPhenoData</span>(
    umi,
    <span class="kw">aes_string</span>(<span class="dt">x =</span> <span class="st">&quot;total_features&quot;</span>,
               <span class="dt">y =</span> <span class="st">&quot;pct_counts_feature_controls_ERCC&quot;</span>,
               <span class="dt">colour =</span> <span class="st">&quot;batch&quot;</span>)
)</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/ercc-vs-counts-1.png" alt="(\#fig:ercc-vs-counts)Percentage of counts in ERCCs" width="90%" />
<p class="caption">
(#fig:ercc-vs-counts)Percentage of counts in ERCCs
</p>
</div>
<p>This analysis shows that majority of the cells from NA19098.r2 batch have a very high ERCC/Endo ratio. Indeed, it has been shown by the authors that this batch contains cells of smaller size.</p>
<p><strong>Exercise</strong></p>
<p>Create filters for removing batch NA19098.r2 and cells with high expression of mitochondrial genes (&gt;10% of total counts in a cell).</p>
<p><strong>Answer</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_ERCC &lt;-<span class="st"> </span>umi$batch !=<span class="st"> &quot;NA19098.r2&quot;</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
  <span class="kw">as.data.frame</span>(<span class="kw">table</span>(filter_by_ERCC)),
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
  <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by ERCC filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-30)The number of cells removed by ERCC filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_ERCC</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">96</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">768</td>
</tr>
</tbody>
</table>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">filter_by_MT &lt;-<span class="st"> </span>umi$pct_counts_feature_controls_MT &lt;<span class="st"> </span><span class="dv">10</span></code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
  <span class="kw">as.data.frame</span>(<span class="kw">table</span>(filter_by_MT)),
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
  <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by MT filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-32)The number of cells removed by MT filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_MT</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">31</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">833</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="cell-filtering" class="section level2">
<h2><span class="header-section-number">6.6</span> Cell filtering</h2>
<div id="default" class="section level3">
<h3><span class="header-section-number">6.6.1</span> Default</h3>
<p>To filter out outlier cells one could use the defaults provided by the scater package:</p>
<ul>
<li><p><strong>filter_on_total_features</strong>: filter out cells based on their total_features being (by default) more than 5 median absolute deviations from the median total_features for the dataset</p></li>
<li><p><strong>filter_on_total_counts</strong>: filter out cells based on their log10-total_counts being (by default) more than 5 median absolute deviations from the median log10-total_counts for the dataset?</p></li>
<li><p><strong>filter_on_pct_counts_feature_controls</strong>: filter out cells on the basis that the percentage of counts from feature controls is higher than a defined threhold (default is 80%)?</p></li>
<li><p><strong>is_cell_control</strong>: filter out cells that were defined as controls</p></li>
</ul>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi$use_default &lt;-<span class="st"> </span>(
    <span class="co"># remove cells with unusual numbers of genes</span>
    !umi$filter_on_total_features &amp;
<span class="st">    </span><span class="co"># sufficient molecules counted</span>
<span class="st">    </span>!umi$filter_on_total_counts &amp;
<span class="st">    </span><span class="co"># sufficient endogenous RNA</span>
<span class="st">    </span>!umi$filter_on_pct_counts_feature_controls_ERCC &amp;
<span class="st">    </span><span class="co"># remove cells with unusual number of reads in MT genes</span>
<span class="st">    </span>!umi$filter_on_pct_counts_feature_controls_MT &amp;
<span class="st">    </span><span class="co"># controls shouldn&#39;t be used in downstream analysis</span>
<span class="st">    </span>!umi$is_cell_control
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
  <span class="kw">as.data.frame</span>(<span class="kw">table</span>(umi$use_default)),
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
  <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by default filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-34)The number of cells removed by default filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">6</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">858</td>
</tr>
</tbody>
</table>
<p>This filtering can be performed even without looking at the plots above.</p>
</div>
<div id="automatic" class="section level3">
<h3><span class="header-section-number">6.6.2</span> Automatic</h3>
<p>Another option available in <strong>scater</strong> is to conduct PCA on a set of QC metrics. The advantage of doing this is that the QC metrics focus on technical aspects of the libraries that are likely to distinguish problematics cells. Automatic outlier detection on PCA plots using QC metrics is available to help identify potentially problematic cells.</p>
<p>By default, the following metrics are used for PCA-based outlier detection:</p>
<ul>
<li><strong>pct_counts_top_100_features</strong></li>
<li><strong>total_features</strong></li>
<li><strong>pct_counts_feature_controls</strong></li>
<li><strong>n_detected_feature_controls</strong></li>
<li><strong>log10_counts_endogenous_features</strong></li>
<li><strong>log10_counts_feature_controls</strong></li>
</ul>
<p>A particular set of variables to be used can be specified with the selected_variables argument as shown in the example below.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi &lt;-
scater::<span class="kw">plotPCA</span>(umi,
                <span class="dt">size_by =</span> <span class="st">&quot;total_features&quot;</span>, 
                <span class="dt">shape_by =</span> <span class="st">&quot;filter_on_total_features&quot;</span>,
                <span class="dt">pca_data_input =</span> <span class="st">&quot;pdata&quot;</span>,
                <span class="dt">detect_outliers =</span> <span class="ot">TRUE</span>,
                <span class="dt">return_SCESet =</span> <span class="ot">TRUE</span>)</code></pre></div>
<pre><code>## The following cells/samples are detected as outliers:
## NA19098.r2.A01
## NA19098.r2.A02
## NA19098.r2.A06
## NA19098.r2.A09
## NA19098.r2.A10
## NA19098.r2.A12
## NA19098.r2.B01
## NA19098.r2.B03
## NA19098.r2.B04
## NA19098.r2.B05
## NA19098.r2.B07
## NA19098.r2.B11
## NA19098.r2.B12
## NA19098.r2.C01
## NA19098.r2.C02
## NA19098.r2.C03
## NA19098.r2.C04
## NA19098.r2.C05
## NA19098.r2.C06
## NA19098.r2.C07
## NA19098.r2.C08
## NA19098.r2.C09
## NA19098.r2.C10
## NA19098.r2.C11
## NA19098.r2.C12
## NA19098.r2.D01
## NA19098.r2.D02
## NA19098.r2.D03
## NA19098.r2.D04
## NA19098.r2.D07
## NA19098.r2.D08
## NA19098.r2.D09
## NA19098.r2.D10
## NA19098.r2.D12
## NA19098.r2.E01
## NA19098.r2.E02
## NA19098.r2.E03
## NA19098.r2.E04
## NA19098.r2.E05
## NA19098.r2.E06
## NA19098.r2.E07
## NA19098.r2.E12
## NA19098.r2.F01
## NA19098.r2.F02
## NA19098.r2.F07
## NA19098.r2.F08
## NA19098.r2.F09
## NA19098.r2.F10
## NA19098.r2.F11
## NA19098.r2.F12
## NA19098.r2.G01
## NA19098.r2.G02
## NA19098.r2.G03
## NA19098.r2.G05
## NA19098.r2.G06
## NA19098.r2.G08
## NA19098.r2.G09
## NA19098.r2.G10
## NA19098.r2.G11
## NA19098.r2.H01
## NA19098.r2.H02
## NA19098.r2.H03
## NA19098.r2.H04
## NA19098.r2.H05
## NA19098.r2.H06
## NA19098.r2.H07
## NA19098.r2.H08
## NA19098.r2.H10
## NA19098.r2.H12
## NA19101.r3.A02
## NA19101.r3.C12
## NA19101.r3.D01
## NA19101.r3.E08
## Variables with highest loadings for PC1 and PC2:
## 
##                                            PC1         PC2
## ---------------------------------  -----------  ----------
## pct_counts_top_100_features          0.4771343   0.3009332
## pct_counts_feature_controls          0.4735839   0.3309562
## n_detected_feature_controls          0.1332811   0.5367629
## log10_counts_feature_controls       -0.1427373   0.5911762
## total_features                      -0.5016681   0.2936705
## log10_counts_endogenous_features    -0.5081855   0.2757918</code></pre>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/auto-cell-filt-1.png" alt="(\#fig:auto-cell-filt)PCA plot used for automatic detection of cell outliers" width="90%" />
<p class="caption">
(#fig:auto-cell-filt)PCA plot used for automatic detection of cell outliers
</p>
</div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
  <span class="kw">as.data.frame</span>(<span class="kw">table</span>(umi$outlier)),
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
  <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by automatic filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-35)The number of cells removed by automatic filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">791</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">73</td>
</tr>
</tbody>
</table>
</div>
<div id="manual" class="section level3">
<h3><span class="header-section-number">6.6.3</span> Manual</h3>
<p>However, since we performed a more detailed analysis, visualized different features and defined our own filters we can use them instead of the default ones:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">umi$use &lt;-<span class="st"> </span>(
    <span class="co"># sufficient features (genes)</span>
    filter_by_expr_features &amp;
<span class="st">    </span><span class="co"># sufficient molecules counted</span>
<span class="st">    </span>filter_by_total_counts &amp;
<span class="st">    </span><span class="co"># sufficient endogenous RNA</span>
<span class="st">    </span>filter_by_ERCC &amp;
<span class="st">    </span><span class="co"># remove cells with unusual number of reads in MT genes</span>
<span class="st">    </span>filter_by_MT
)</code></pre></div>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr::<span class="kw">kable</span>(
  <span class="kw">as.data.frame</span>(<span class="kw">table</span>(umi$use)),
  <span class="dt">booktabs =</span> <span class="ot">TRUE</span>,
  <span class="dt">row.names =</span> <span class="ot">FALSE</span>,
  <span class="dt">caption =</span> <span class="st">&#39;The number of cells removed by manual filter (FALSE)&#39;</span>
)</code></pre></div>
<table>
<caption>(#tab:unnamed-chunk-37)The number of cells removed by manual filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">141</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">723</td>
</tr>
</tbody>
</table>
</div>
</div>
<div id="compare-filterings" class="section level2">
<h2><span class="header-section-number">6.7</span> Compare filterings</h2>
<p><strong>Exercise</strong></p>
<p>Compare the default, automatic and manual cell filters. Plot a Venn diagram of the outlier cells from these filterings.</p>
<p><strong>Hint</strong>: Use <code>limma::vennCounts</code> and <code>limma::vennDiagram</code> functions from the <a href="https://bioconductor.org/packages/release/bioc/html/limma.html">limma</a> package to make a Venn diagram.</p>
<p><strong>Answer</strong></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">def &lt;-<span class="st"> </span><span class="kw">colnames</span>(umi)[!umi$use_default]
auto &lt;-<span class="st"> </span><span class="kw">colnames</span>(umi)[umi$outlier]
man &lt;-<span class="st"> </span><span class="kw">colnames</span>(umi)[!umi$use]
venn.diag &lt;-<span class="st"> </span>limma::<span class="kw">vennCounts</span>(<span class="kw">cbind</span>(<span class="kw">colnames</span>(umi) %in%<span class="st"> </span>def,
                                     <span class="kw">colnames</span>(umi) %in%<span class="st"> </span>auto,
                                     <span class="kw">colnames</span>(umi) %in%<span class="st"> </span>man))
limma::<span class="kw">vennDiagram</span>(venn.diag,
                   <span class="dt">names =</span> <span class="kw">c</span>(<span class="st">&quot;Default&quot;</span>, <span class="st">&quot;Automatic&quot;</span>, <span class="st">&quot;Manual&quot;</span>),
                   <span class="dt">circle.col =</span> <span class="kw">c</span>(<span class="st">&quot;magenta&quot;</span>, <span class="st">&quot;blue&quot;</span>, <span class="st">&quot;green&quot;</span>))</code></pre></div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/cell-filt-comp-1.png" alt="(\#fig:cell-filt-comp)Comparison of the default, automatic and manual cell filters" width="90%" />
<p class="caption">
(#fig:cell-filt-comp)Comparison of the default, automatic and manual cell filters
</p>
</div>
</div>
<div id="save-the-data" class="section level2">
<h2><span class="header-section-number">6.8</span> Save the data</h2>
<p>Dimensions of the QCed dataset (do not forget about the gene filter we defined above):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">dim</span>(umi[<span class="kw">fData</span>(umi)$use, <span class="kw">pData</span>(umi)$use])</code></pre></div>
<pre><code>## Features  Samples 
##    14214      723</code></pre>
<p>Save the data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">saveRDS</span>(umi, <span class="dt">file =</span> <span class="st">&quot;blischak/umi.rds&quot;</span>)</code></pre></div>
</div>
<div id="exercise" class="section level2">
<h2><span class="header-section-number">6.9</span> Exercise</h2>
<p>Perform exactly the same QC analysis with read counts of the same Blischak data. Use <code>blischak/reads.txt</code> file to load the reads. Once you have finished please compare your results to ours (next chapter).</p>
<!--chapter:end:05-exprs-qc.Rmd-->
</div>
</div>
<div id="expression-qc-reads" class="section level1">
<h1><span class="header-section-number">7</span> Expression QC (Reads)</h1>
<table>
<caption>(#tab:unnamed-chunk-42)A table of the first 6 rows and 3 columns of the molecules table.</caption>
<thead>
<tr class="header">
<th></th>
<th align="right">NA19098.r1.A01</th>
<th align="right">NA19098.r1.A02</th>
<th align="right">NA19098.r1.A03</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>ENSG00000237683</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>ENSG00000187634</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>ENSG00000188976</td>
<td align="right">57</td>
<td align="right">140</td>
<td align="right">1</td>
</tr>
<tr class="even">
<td>ENSG00000187961</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="odd">
<td>ENSG00000187583</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
<tr class="even">
<td>ENSG00000187642</td>
<td align="right">0</td>
<td align="right">0</td>
<td align="right">0</td>
</tr>
</tbody>
</table>
<table>
<caption>(#tab:unnamed-chunk-42)A table of the first 6 rows of the anno table.</caption>
<thead>
<tr class="header">
<th align="left">individual</th>
<th align="left">replicate</th>
<th align="left">well</th>
<th align="left">batch</th>
<th align="left">sample_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A01</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A01</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A02</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A02</td>
</tr>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A03</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A03</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A04</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A04</td>
</tr>
<tr class="odd">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A05</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A05</td>
</tr>
<tr class="even">
<td align="left">NA19098</td>
<td align="left">r1</td>
<td align="left">A06</td>
<td align="left">NA19098.r1</td>
<td align="left">NA19098.r1.A06</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/dropout-overview-reads-1.png" alt="(\#fig:dropout-overview-reads)Dropout rate vs mean expression" width="90%" />
<p class="caption">
(#fig:dropout-overview-reads)Dropout rate vs mean expression
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/top50-gene-expr-reads-1.png" alt="(\#fig:top50-gene-expr-reads)Number of total counts consumed by the top 50 expressed genes" width="90%" />
<p class="caption">
(#fig:top50-gene-expr-reads)Number of total counts consumed by the top 50 expressed genes
</p>
</div>
<table>
<caption>(#tab:unnamed-chunk-48)The number of genes removed by gene filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_genes</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">2445</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">16281</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-counts-hist-reads-1.png" alt="(\#fig:total-counts-hist-reads)Histogram of library sizes for all cells" width="90%" />
<p class="caption">
(#fig:total-counts-hist-reads)Histogram of library sizes for all cells
</p>
</div>
<table>
<caption>(#tab:unnamed-chunk-50)The number of cells removed by total counts filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_total_counts</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">180</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">684</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-features-hist-reads-1.png" alt="(\#fig:total-features-hist-reads)Histogram of the number of detected genes in all cells" width="90%" />
<p class="caption">
(#fig:total-features-hist-reads)Histogram of the number of detected genes in all cells
</p>
</div>
<table>
<caption>(#tab:unnamed-chunk-52)The number of cells removed by total features filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_expr_features</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">16</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">848</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/total-features-vs-counts-reads-1.png" alt="(\#fig:total-features-vs-counts-reads)Library size vs number of detected genes" width="90%" />
<p class="caption">
(#fig:total-features-vs-counts-reads)Library size vs number of detected genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/mt-vs-counts-reads-1.png" alt="(\#fig:mt-vs-counts-reads)Percentage of counts in MT genes" width="90%" />
<p class="caption">
(#fig:mt-vs-counts-reads)Percentage of counts in MT genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/ercc-vs-counts-reads-1.png" alt="(\#fig:ercc-vs-counts-reads)Percentage of counts in ERCCs" width="90%" />
<p class="caption">
(#fig:ercc-vs-counts-reads)Percentage of counts in ERCCs
</p>
</div>
<table>
<caption>(#tab:unnamed-chunk-54)The number of cells removed by ERCC filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_ERCC</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">103</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">761</td>
</tr>
</tbody>
</table>
<table>
<caption>(#tab:unnamed-chunk-56)The number of cells removed by MT filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">filter_by_MT</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">18</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">846</td>
</tr>
</tbody>
</table>
<table>
<caption>(#tab:unnamed-chunk-58)The number of cells removed by default filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">37</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">827</td>
</tr>
</tbody>
</table>
<pre><code>## The following cells/samples are detected as outliers:
## NA19098.r1.B10
## NA19098.r1.D07
## NA19098.r1.E04
## NA19098.r1.F06
## NA19098.r1.H08
## NA19098.r1.H09
## NA19098.r2.A01
## NA19098.r2.A06
## NA19098.r2.A09
## NA19098.r2.A12
## NA19098.r2.B01
## NA19098.r2.B11
## NA19098.r2.B12
## NA19098.r2.C04
## NA19098.r2.C09
## NA19098.r2.D02
## NA19098.r2.D03
## NA19098.r2.D09
## NA19098.r2.E04
## NA19098.r2.E07
## NA19098.r2.F01
## NA19098.r2.F11
## NA19098.r2.G01
## NA19098.r2.G05
## NA19098.r2.G10
## NA19098.r2.H01
## NA19098.r2.H07
## NA19098.r2.H08
## NA19098.r2.H12
## NA19098.r3.A05
## NA19098.r3.A07
## NA19098.r3.B02
## NA19098.r3.C07
## NA19098.r3.E05
## NA19098.r3.E08
## NA19098.r3.E09
## NA19098.r3.F11
## NA19098.r3.F12
## NA19098.r3.G02
## NA19098.r3.G03
## NA19098.r3.G04
## NA19098.r3.G11
## NA19098.r3.G12
## NA19098.r3.H08
## NA19101.r1.A01
## NA19101.r1.A12
## NA19101.r1.B01
## NA19101.r1.B06
## NA19101.r1.E09
## NA19101.r1.E11
## NA19101.r1.F05
## NA19101.r1.F10
## NA19101.r1.G01
## NA19101.r1.G06
## NA19101.r1.H04
## NA19101.r1.H09
## NA19101.r2.A03
## NA19101.r2.C10
## NA19101.r2.E05
## NA19101.r2.F02
## NA19101.r2.H04
## NA19101.r2.H10
## NA19101.r3.A02
## NA19101.r3.A03
## NA19101.r3.A05
## NA19101.r3.A09
## NA19101.r3.B05
## NA19101.r3.C01
## NA19101.r3.C09
## NA19101.r3.C12
## NA19101.r3.D01
## NA19101.r3.D04
## NA19101.r3.D07
## NA19101.r3.D09
## NA19101.r3.E08
## NA19101.r3.F09
## NA19101.r3.G09
## NA19101.r3.H01
## NA19101.r3.H03
## NA19101.r3.H07
## NA19101.r3.H09
## NA19239.r1.F05
## NA19239.r1.G05
## NA19239.r2.B01
## NA19239.r2.B03
## NA19239.r2.B10
## NA19239.r2.B11
## NA19239.r2.C03
## NA19239.r2.C06
## NA19239.r2.C08
## NA19239.r2.D07
## NA19239.r2.D09
## NA19239.r2.E09
## NA19239.r2.F04
## NA19239.r2.F06
## NA19239.r2.F07
## NA19239.r2.F12
## NA19239.r2.G03
## NA19239.r2.G08
## NA19239.r2.H02
## NA19239.r2.H03
## NA19239.r2.H07
## NA19239.r3.A01
## NA19239.r3.B09
## NA19239.r3.C04
## NA19239.r3.C07
## NA19239.r3.E01
## NA19239.r3.E03
## NA19239.r3.E12
## NA19239.r3.H02
## NA19239.r3.H10
## Variables with highest loadings for PC1 and PC2:
## 
##                                            PC1         PC2
## ---------------------------------  -----------  ----------
## pct_counts_feature_controls          0.5057646   0.2473134
## pct_counts_top_100_features          0.4888852   0.2277068
## n_detected_feature_controls          0.0231277   0.6235516
## log10_counts_feature_controls       -0.1226860   0.6576822
## total_features                      -0.4655518   0.2219694
## log10_counts_endogenous_features    -0.5223679   0.1278782</code></pre>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/auto-cell-filt-reads-1.png" alt="(\#fig:auto-cell-filt-reads)PCA plot used for automatic detection of cell outliers" width="90%" />
<p class="caption">
(#fig:auto-cell-filt-reads)PCA plot used for automatic detection of cell outliers
</p>
</div>
<table>
<caption>(#tab:unnamed-chunk-59)The number of cells removed by automatic filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">753</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">111</td>
</tr>
</tbody>
</table>
<table>
<caption>(#tab:unnamed-chunk-61)The number of cells removed by manual filter (FALSE)</caption>
<thead>
<tr class="header">
<th align="left">Var1</th>
<th align="right">Freq</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td align="left">FALSE</td>
<td align="right">254</td>
</tr>
<tr class="even">
<td align="left">TRUE</td>
<td align="right">610</td>
</tr>
</tbody>
</table>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/cell-filt-comp-reads-1.png" alt="(\#fig:cell-filt-comp-reads)Comparison of the default, automatic and manual cell filters" width="90%" />
<p class="caption">
(#fig:cell-filt-comp-reads)Comparison of the default, automatic and manual cell filters
</p>
</div>
<pre><code>## Features  Samples 
##    16281      610</code></pre>
<!--chapter:end:06-exprs-qc-reads.Rmd-->
</div>
<div id="expression-overview" class="section level1">
<h1><span class="header-section-number">8</span> Expression overview</h1>
<div id="introduction-2" class="section level2">
<h2><span class="header-section-number">8.1</span> Introduction</h2>
<p>Here we will continue to work with the filtered <strong>blischak</strong> dataset produced in the previous chapter. We will look at what happened to the expression matrix after the quality control step.</p>
</div>
<div id="top-500-genes" class="section level2">
<h2><span class="header-section-number">8.2</span> Top 500 genes</h2>
<p>With scater one can look at the proportion of reads accounted by the top 500 expressed genes.</p>
<div id="before-qc" class="section level3">
<h3><span class="header-section-number">8.2.1</span> Before QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-top500-before-qc-1.png" alt="(\#fig:expr-overview-top500-before-qc)Proportion of reads accounted by the top 500" width="90%" />
<p class="caption">
(#fig:expr-overview-top500-before-qc)Proportion of reads accounted by the top 500
</p>
</div>
</div>
<div id="after-qc" class="section level3">
<h3><span class="header-section-number">8.2.2</span> After QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-top500-after-qc-1.png" alt="(\#fig:expr-overview-top500-after-qc)Proportion of reads accounted by the top 500" width="90%" />
<p class="caption">
(#fig:expr-overview-top500-after-qc)Proportion of reads accounted by the top 500
</p>
</div>
</div>
</div>
<div id="pca-plot" class="section level2">
<h2><span class="header-section-number">8.3</span> PCA plot</h2>
<p>The easiest thing to overview the data is to transform it using the principal component analysis and then visualize the first two principal components. Again the <a href="https://github.com/davismcc/scater">scater</a> package provides several very useful functions to make this analysis straightforward.</p>
<div id="before-qc-1" class="section level3">
<h3><span class="header-section-number">8.3.1</span> Before QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-pca-before-qc-1.png" alt="(\#fig:expr-overview-pca-before-qc)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-pca-before-qc)PCA plot of the blischak data
</p>
</div>
</div>
<div id="after-qc-1" class="section level3">
<h3><span class="header-section-number">8.3.2</span> After QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-pca-after-qc-1.png" alt="(\#fig:expr-overview-pca-after-qc)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-pca-after-qc)PCA plot of the blischak data
</p>
</div>
</div>
</div>
<div id="diffusion-map" class="section level2">
<h2><span class="header-section-number">8.4</span> Diffusion map</h2>
<p>Another way of representing the data is a diffusion map. It is very useful if the cells represent a continuous process (e.g. development).</p>
<div id="before-qc-2" class="section level3">
<h3><span class="header-section-number">8.4.1</span> Before QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-diff-before-qc-1.png" alt="(\#fig:expr-overview-diff-before-qc)Diffusion map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-diff-before-qc)Diffusion map of the blischak data
</p>
</div>
</div>
<div id="after-qc-2" class="section level3">
<h3><span class="header-section-number">8.4.2</span> After QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-diff-after-qc-1.png" alt="(\#fig:expr-overview-diff-after-qc)Diffusion map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-diff-after-qc)Diffusion map of the blischak data
</p>
</div>
</div>
</div>
<div id="tsne-map" class="section level2">
<h2><span class="header-section-number">8.5</span> tSNE map</h2>
<p>Another way of representing the data is a tSNE map.</p>
<div id="before-qc-3" class="section level3">
<h3><span class="header-section-number">8.5.1</span> Before QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-tsne-before-qc-1.png" alt="(\#fig:expr-overview-tsne-before-qc)tSNE map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-tsne-before-qc)tSNE map of the blischak data
</p>
</div>
</div>
<div id="after-qc-3" class="section level3">
<h3><span class="header-section-number">8.5.2</span> After QC</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-tsne-after-qc-1.png" alt="(\#fig:expr-overview-tsne-after-qc)tSNE map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-tsne-after-qc)tSNE map of the blischak data
</p>
</div>
</div>
</div>
<div id="exercise-1" class="section level2">
<h2><span class="header-section-number">8.6</span> Exercise</h2>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>blischak/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter).</p>
<!--chapter:end:07-exprs-overview.Rmd-->
</div>
</div>
<div id="expression-overview-reads" class="section level1">
<h1><span class="header-section-number">9</span> Expression overview (Reads)</h1>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-top500-before-qc-reads-1.png" alt="(\#fig:expr-overview-top500-before-qc-reads)Proportion of reads accounted by the top 500" width="90%" />
<p class="caption">
(#fig:expr-overview-top500-before-qc-reads)Proportion of reads accounted by the top 500
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-top500-after-qc-reads-1.png" alt="(\#fig:expr-overview-top500-after-qc-reads)Proportion of reads accounted by the top 500" width="90%" />
<p class="caption">
(#fig:expr-overview-top500-after-qc-reads)Proportion of reads accounted by the top 500
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-pca-before-qc-reads-1.png" alt="(\#fig:expr-overview-pca-before-qc-reads)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-pca-before-qc-reads)PCA plot of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-pca-after-qc-reads-1.png" alt="(\#fig:expr-overview-pca-after-qc-reads)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-pca-after-qc-reads)PCA plot of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-diff-before-qc-reads-1.png" alt="(\#fig:expr-overview-diff-before-qc-reads)Diffusion map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-diff-before-qc-reads)Diffusion map of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-diff-after-qc-reads-1.png" alt="(\#fig:expr-overview-diff-after-qc-reads)Diffusion map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-diff-after-qc-reads)Diffusion map of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-tsne-before-qc-reads-1.png" alt="(\#fig:expr-overview-tsne-before-qc-reads)tSNE map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-tsne-before-qc-reads)tSNE map of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/expr-overview-tsne-after-qc-reads-1.png" alt="(\#fig:expr-overview-tsne-after-qc-reads)tSNE map of the blischak data" width="90%" />
<p class="caption">
(#fig:expr-overview-tsne-after-qc-reads)tSNE map of the blischak data
</p>
</div>
<!--chapter:end:08-exprs-overview-reads.Rmd-->
</div>
<div id="confounding-factors" class="section level1">
<h1><span class="header-section-number">10</span> Confounding factors</h1>
<div id="introduction-3" class="section level2">
<h2><span class="header-section-number">10.1</span> Introduction</h2>
<p>There are a very large number of potential confounders, artifacts and biases in sc-RNA-seq data One of the main challenges in analyzing scRNA-seq data stems from the fact that it is difficult to carry out a true technical replicate (why?) to distinguish biological and technical variability. Exploring the effects of such variabilities (both those recorded during the experiment and computed QC metrics) is crucial for appropriate modeling of the data.</p>
<p>The <a href="https://github.com/davismcc/scater">scater</a> package provides a set of methods specifically for quality control of experimental and explanatory variables.</p>
<p>Here we will continue to work with the Blischak data that was used in the previous chapter.</p>
<p>The <code>umi.qc</code> dataset contains filtered cells and genes. Our next step is to explore technical drivers of variability in the data to inform data normalisation before downstream analysis.</p>
</div>
<div id="correlations-with-pcs" class="section level2">
<h2><span class="header-section-number">10.2</span> Correlations with PCs</h2>
Let’s first look again at the PCA plot of the QCed dataset:
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-pca-1.png" alt="(\#fig:confound-pca)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:confound-pca)PCA plot of the blischak data
</p>
</div>
<p>scater allows one to identify principle components that correlate with experimental and QC variables of interest (it ranks principle components by <span class="math inline">\(R^2\)</span> from a linear model regressing PC value against the variable of interest).</p>
<p>Let’s test whether some of the variables correlate with any of the PCs.</p>
<div id="detected-genes" class="section level3">
<h3><span class="header-section-number">10.2.1</span> Detected genes</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-total-features-1.png" alt="(\#fig:confound-find-pcs-total-features)PC correlation with the number of detected genes" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-total-features)PC correlation with the number of detected genes
</p>
</div>
<p>Indeed, we can see that PC1 can be completely explained by the number of the detected genes. In fact, it was also visible on the PCA plot above. This is a well-known issue in scRNA-seq and was described <a href="http://biorxiv.org/content/early/2015/12/27/025528">here</a>.</p>
</div>
<div id="control-erccs" class="section level3">
<h3><span class="header-section-number">10.2.2</span> Control ERCCs</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-erccs-1.png" alt="(\#fig:confound-find-pcs-erccs)PC correlation with the percentage of reads in ERCCs" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-erccs)PC correlation with the percentage of reads in ERCCs
</p>
</div>
</div>
<div id="control-mts" class="section level3">
<h3><span class="header-section-number">10.2.3</span> Control MTs</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-mts-1.png" alt="(\#fig:confound-find-pcs-mts)PC correlation with the percentage of reads in MT genes" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-mts)PC correlation with the percentage of reads in MT genes
</p>
</div>
</div>
</div>
<div id="explanatory-variables" class="section level2">
<h2><span class="header-section-number">10.3</span> Explanatory variables</h2>
<p>scater can also compute the marginal <span class="math inline">\(R^2\)</span> for each variable when fitting a linear model regressing expression values for each gene against just that variable, and display a density plot of the gene-wise marginal <span class="math inline">\(R^2\)</span> values for the variables.</p>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-expl-vars-1.png" alt="(\#fig:confound-find-expl-vars)Explanatory variables" width="90%" />
<p class="caption">
(#fig:confound-find-expl-vars)Explanatory variables
</p>
</div>
<p>This analysis indicates that the number of detected genes (again) and also the sequencing depth (number of counts) have substantial explanatory power for many genes, so these variables are good candidates for conditioning out in a normalisation step, or including in downstream statistical models. Expression of ERCCs also appears to be an important explanatory variable.</p>
</div>
<div id="other-confounders" class="section level2">
<h2><span class="header-section-number">10.4</span> Other confounders</h2>
<p>In addition to correcting for batch, there are other factors that one may want to compensate for. As with batch correction, these adjustments require extrinsic information. One popular method is <a href="https://github.com/PMBio/scLVM">scLVM</a> which allows you to identify and subtract the effect from processes such as cell-cycle or apoptosis.</p>
</div>
<div id="exercise-2" class="section level2">
<h2><span class="header-section-number">10.5</span> Exercise</h2>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>blischak/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter).</p>
<!--chapter:end:09-confounders.Rmd-->
</div>
</div>
<div id="confounding-factors-reads" class="section level1">
<h1><span class="header-section-number">11</span> Confounding factors (Reads)</h1>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-pca-reads-1.png" alt="(\#fig:confound-pca-reads)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:confound-pca-reads)PCA plot of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-total-features-reads-1.png" alt="(\#fig:confound-find-pcs-total-features-reads)PC correlation with the number of detected genes" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-total-features-reads)PC correlation with the number of detected genes
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-erccs-reads-1.png" alt="(\#fig:confound-find-pcs-erccs-reads)PC correlation with the percentage of reads in ERCCs" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-erccs-reads)PC correlation with the percentage of reads in ERCCs
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/confound-find-pcs-mts-reads-1.png" alt="(\#fig:confound-find-pcs-mts-reads)PC correlation with the percentage of reads in MT genes" width="90%" />
<p class="caption">
(#fig:confound-find-pcs-mts-reads)PC correlation with the percentage of reads in MT genes
</p>
</div>
<!--chapter:end:10-confounders-reads.Rmd-->
</div>
<div id="normalization-for-library-size" class="section level1">
<h1><span class="header-section-number">12</span> Normalization for library size</h1>
<div id="introduction-4" class="section level2">
<h2><span class="header-section-number">12.1</span> Introduction</h2>
<p>In the previous chapter we identified important confounding factors and explanatory variables. scater allows one to account for these variables in subsequent statistical models or to condition them out using <code>normaliseExprs()</code>, if so desired. This can be done by providing a design matrix to <code>normaliseExprs()</code>. We are not covering this topic here, but you can try to do it yourself as an exercise.</p>
<p>Instead we will explore how simple size-factor normalisations correcting for library size can remove the effects of some of the confounders and explanatory variables.</p>
</div>
<div id="library-size-1" class="section level2">
<h2><span class="header-section-number">12.2</span> Library size</h2>
<p>Library sizes vary because of the various reasons: * scRNA-seq data is often sequenced on highly multiplexed platforms the total reads which are derived from each cell may differ substantially. * Protocols may differ in terms of their coverage of each transcript, their bias based on the average content of <strong>A/T</strong> nucleotides, or their ability to capture short transcripts.</p>
<p>Ideally, we would like to compensate for all of these differences and biases when comparing data from two different experiments.</p>
<p>Many methods to correct for library size have been developped for bulk RNA-seq and can be equally applied to scRNA-seq (eg. UQ, SF, CPM, RPKM, FPKM, TPM). Some quantification methods (eg. <a href="http://cole-trapnell-lab.github.io/cufflinks/">Cufflinks</a>, <a href="http://deweylab.github.io/RSEM/">RSEM</a>) incorporated library size when determining gene expression estimates thus do not require this normalization.</p>
<p>We will continue to work with the Blischak data that was used in the previous chapter and show how scater makes it easy to perform different types of size-factor normalizations.</p>
</div>
<div id="normalisations" class="section level2">
<h2><span class="header-section-number">12.3</span> Normalisations</h2>
<p>The simplest way to normalize this data is to convert it to counts per million (<strong>CPM</strong>) by dividing each column by its total then multiplying by 1,000,000. Note that spike-ins should be excluded from the calculation of total expression in order to correct for total cell RNA content, therefore we will only use endogenous genes. scater performs this normalisation by default, you can control it by changing <code>exprs_values</code> parameter to <code>&quot;exprs&quot;</code>.</p>
<p>Another method is called <strong>TMM</strong> is the weighted trimmed mean of M-values (to the reference) proposed by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-3-r25">Robinson and Oshlack (2010)</a>, where the weights are from the delta method on Binomial data.</p>
<p>Another very popular method <strong>RLE</strong> is the scaling factor method proposed by <a href="https://genomebiology.biomedcentral.com/articles/10.1186/gb-2010-11-10-r106">Anders and Huber (2010)</a>. We call it “relative log expression”, as median library is calculated from the geometric mean of all columns and the median ratio of each sample to the median library is taken as the scale factor.</p>
<p>A similar to <strong>RLE</strong> the <strong>upperquartile</strong> is the upper-quartile normalization method of <a href="http://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-11-94">Bullard et al (2010)</a>, in which the scale factors are calculated from the 75% quantile of the counts for each library, after removing genes which are zero in all libraries. This idea is generalized here to allow scaling by any quantile of the distributions.</p>
<p>Let’s compare all these methods.</p>
<div id="raw" class="section level3">
<h3><span class="header-section-number">12.3.1</span> Raw</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-raw-1.png" alt="(\#fig:norm-pca-raw)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:norm-pca-raw)PCA plot of the blischak data
</p>
</div>
</div>
<div id="cpm" class="section level3">
<h3><span class="header-section-number">12.3.2</span> CPM</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-cpm-1.png" alt="(\#fig:norm-pca-cpm)PCA plot of the blischak data after CPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-cpm)PCA plot of the blischak data after CPM normalisation
</p>
</div>
</div>
<div id="log2cpm" class="section level3">
<h3><span class="header-section-number">12.3.3</span> log2(CPM)</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-log2-cpm-1.png" alt="(\#fig:norm-pca-log2-cpm)PCA plot of the blischak data after log2(CPM) normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-log2-cpm)PCA plot of the blischak data after log2(CPM) normalisation
</p>
</div>
</div>
<div id="tmm" class="section level3">
<h3><span class="header-section-number">12.3.4</span> TMM</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-tmm-1.png" alt="(\#fig:norm-pca-tmm)PCA plot of the blischak data after TMM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-tmm)PCA plot of the blischak data after TMM normalisation
</p>
</div>
</div>
<div id="rle" class="section level3">
<h3><span class="header-section-number">12.3.5</span> RLE</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-rle-1.png" alt="(\#fig:norm-pca-rle)PCA plot of the blischak data after RLE normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-rle)PCA plot of the blischak data after RLE normalisation
</p>
</div>
</div>
<div id="upperquantile" class="section level3">
<h3><span class="header-section-number">12.3.6</span> Upperquantile</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-uq-1.png" alt="(\#fig:norm-pca-uq)PCA plot of the blischak data after UQ normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-uq)PCA plot of the blischak data after UQ normalisation
</p>
</div>
</div>
</div>
<div id="interpretation" class="section level2">
<h2><span class="header-section-number">12.4</span> Interpretation</h2>
<p>Clearly, only the CPM normalisation has reduced the effect of the number of detected genes and separated cells by individuals (though very weakly).</p>
</div>
<div id="other-methods" class="section level2">
<h2><span class="header-section-number">12.5</span> Other methods</h2>
<p>Some methods combine library size and fragment/gene length normalization such as:</p>
<ul>
<li><strong>RPKM</strong> - Reads Per Kilobase Million (for single-end sequencing)</li>
<li><strong>FPKM</strong> - Fragments Per Kilobase Million (same as <strong>RPKM</strong> but for paired-end sequencing, makes sure that paired ends mapped to the same fragment are not counted twice)</li>
<li><strong>TPM</strong> - Transcripts Per Kilobase Million (same as <strong>RPKM</strong>, but the order of normalizations is reversed - length first and sequencing depth second)</li>
</ul>
<p>These methods are not applicable to our dataset since the end of the transcript which contains the UMI was preferentially sequenced. Furthermore in general these should only be calculated using appropriate quantification software from aligned BAM files not from read counts since often only a portion of the entire gene/transcript is sequenced, not the entire length.</p>
<p>However, here we show how these normalisations can be calculated using scater. First, we need to find the effective transcript length in Kilobases. However, our dataset containes only gene IDs, therefore we will be using the gene lengths instead of transcripts. scater uses the <a href="https://bioconductor.org/packages/release/bioc/html/biomaRt.html">biomaRt</a> package, which allows one to annotate genes by other attributes:</p>
<p>Some of the genes were not annotated, therefore we filter them out:</p>
<p>Now we compute the effective gene length in Kilobases by using the <code>end_position</code> and <code>start_position</code> fields:</p>
<p>Now we are ready to perform the normalisations:</p>
Plot the results as a PCA plot:
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-fpkm-1.png" alt="(\#fig:norm-pca-fpkm)PCA plot of the blischak data after FPKM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-fpkm)PCA plot of the blischak data after FPKM normalisation
</p>
</div>
<p>TPM normalisation produce a zero-matrix, we are not sure why, it maybe a bug in scater.</p>
</div>
<div id="visualize-genes" class="section level2">
<h2><span class="header-section-number">12.6</span> Visualize genes</h2>
<p>Now after the normalisation we are ready to visualise the gene expression:</p>
<div id="raw-1" class="section level3">
<h3><span class="header-section-number">12.6.1</span> Raw</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-raw-1.png" alt="(\#fig:norm-genes-raw)Expression of the first 6 genes of the blischak data" width="90%" />
<p class="caption">
(#fig:norm-genes-raw)Expression of the first 6 genes of the blischak data
</p>
</div>
</div>
<div id="cpm-1" class="section level3">
<h3><span class="header-section-number">12.6.2</span> CPM</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-cpm-1.png" alt="(\#fig:norm-genes-cpm)Expression of the first 6 genes of the blischak data after the CPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-cpm)Expression of the first 6 genes of the blischak data after the CPM normalisation
</p>
</div>
</div>
<div id="log2cpm-1" class="section level3">
<h3><span class="header-section-number">12.6.3</span> log2(CPM)</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-log2-cpm-1.png" alt="(\#fig:norm-genes-log2-cpm)Expression of the first 6 genes of the blischak data after the log2(CPM) normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-log2-cpm)Expression of the first 6 genes of the blischak data after the log2(CPM) normalisation
</p>
</div>
</div>
<div id="upperquantile-1" class="section level3">
<h3><span class="header-section-number">12.6.4</span> Upperquantile</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-UQ-1.png" alt="(\#fig:norm-genes-UQ)Expression of the first 6 genes of the blischak data after the UQ normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-UQ)Expression of the first 6 genes of the blischak data after the UQ normalisation
</p>
</div>
</div>
<div id="fpkm" class="section level3">
<h3><span class="header-section-number">12.6.5</span> FPKM</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-fpkm-1.png" alt="(\#fig:norm-genes-fpkm)Expression of the first 6 genes of the blischak data after the FPKM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-fpkm)Expression of the first 6 genes of the blischak data after the FPKM normalisation
</p>
</div>
</div>
<div id="tpm" class="section level3">
<h3><span class="header-section-number">12.6.6</span> TPM</h3>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-tpm-1.png" alt="(\#fig:norm-genes-tpm)Expression of the first 6 genes of the blischak data after the TPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-tpm)Expression of the first 6 genes of the blischak data after the TPM normalisation
</p>
</div>
</div>
</div>
<div id="exercise-3" class="section level2">
<h2><span class="header-section-number">12.7</span> Exercise</h2>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>blischak/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter).</p>
<!--chapter:end:11-exprs-norm.Rmd-->
</div>
</div>
<div id="normalization-for-library-size-reads" class="section level1">
<h1><span class="header-section-number">13</span> Normalization for library size (Reads)</h1>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-raw-reads-1.png" alt="(\#fig:norm-pca-raw-reads)PCA plot of the blischak data" width="90%" />
<p class="caption">
(#fig:norm-pca-raw-reads)PCA plot of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-cpm-reads-1.png" alt="(\#fig:norm-pca-cpm-reads)PCA plot of the blischak data after CPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-cpm-reads)PCA plot of the blischak data after CPM normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-log2-cpm-reads-1.png" alt="(\#fig:norm-pca-log2-cpm-reads)PCA plot of the blischak data after log2(CPM) normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-log2-cpm-reads)PCA plot of the blischak data after log2(CPM) normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-tmm-reads-1.png" alt="(\#fig:norm-pca-tmm-reads)PCA plot of the blischak data after TMM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-tmm-reads)PCA plot of the blischak data after TMM normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-rle-reads-1.png" alt="(\#fig:norm-pca-rle-reads)PCA plot of the blischak data after RLE normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-rle-reads)PCA plot of the blischak data after RLE normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-uq-reads-1.png" alt="(\#fig:norm-pca-uq-reads)PCA plot of the blischak data after UQ normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-uq-reads)PCA plot of the blischak data after UQ normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-pca-fpkm-reads-1.png" alt="(\#fig:norm-pca-fpkm-reads)PCA plot of the blischak data after FPKM normalisation" width="90%" />
<p class="caption">
(#fig:norm-pca-fpkm-reads)PCA plot of the blischak data after FPKM normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-raw-reads-1.png" alt="(\#fig:norm-genes-raw-reads)Expression of the first 6 genes of the blischak data" width="90%" />
<p class="caption">
(#fig:norm-genes-raw-reads)Expression of the first 6 genes of the blischak data
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-cpm-reads-1.png" alt="(\#fig:norm-genes-cpm-reads)Expression of the first 6 genes of the blischak data after the CPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-cpm-reads)Expression of the first 6 genes of the blischak data after the CPM normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-log2-cpm-reads-1.png" alt="(\#fig:norm-genes-log2-cpm-reads)Expression of the first 6 genes of the blischak data after the log2(CPM) normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-log2-cpm-reads)Expression of the first 6 genes of the blischak data after the log2(CPM) normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-UQ-reads-1.png" alt="(\#fig:norm-genes-UQ-reads)Expression of the first 6 genes of the blischak data after the UQ normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-UQ-reads)Expression of the first 6 genes of the blischak data after the UQ normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-fpkm-reads-1.png" alt="(\#fig:norm-genes-fpkm-reads)Expression of the first 6 genes of the blischak data after the FPKM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-fpkm-reads)Expression of the first 6 genes of the blischak data after the FPKM normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/norm-genes-tpm-reads-1.png" alt="(\#fig:norm-genes-tpm-reads)Expression of the first 6 genes of the blischak data after the TPM normalisation" width="90%" />
<p class="caption">
(#fig:norm-genes-tpm-reads)Expression of the first 6 genes of the blischak data after the TPM normalisation
</p>
</div>
<!--chapter:end:12-exprs-norm-reads.Rmd-->
</div>
<div id="remove-confounders-using-controls" class="section level1">
<h1><span class="header-section-number">14</span> Remove confounders using controls</h1>
<div id="introduction-5" class="section level2">
<h2><span class="header-section-number">14.1</span> Introduction</h2>
<p>In the previous chapter we saw that the library size normalisation is able to remove major confounders. Let us see whether further use of the controls (ERCC and MT) can provide us with more insights into the data.</p>
<p>Several methods (eg. <a href="https://github.com/catavallejos/BASiCS">BASiCS</a>, <a href="https://github.com/PMBio/scLVM">scLVM</a>, <a href="http://bioconductor.org/packages/release/bioc/html/RUVSeq.html">RUV</a>) have been developed for normalisation using ERCCs using different noise models and different fitting procedures.</p>
<p>We will demonstrate some of the methods starting from the simplest one proposed by <a href="http://www.nature.com/nmeth/journal/v10/n11/full/nmeth.2645.html">Brennecke et al.</a>, which identifies genes with significant variation above technical noise (ERCCs).</p>
</div>
<div id="brennecke-method" class="section level2">
<h2><span class="header-section-number">14.2</span> Brennecke method</h2>
<p>To use the method, we first normalize for library size then calculate the mean and the square coefficient of variation (variation divided by the squared mean expression). A curve is fit for the relationship between these two variables for the ERCC spike-in (subject to just technical variation) then a chi-square test is used to find genes significantly above the curve. This has been provided for you as the Brenneck_getVariableGenes(counts, spikes) function.</p>
<pre><code>## Warning in scRNA.seq.funcs::Brennecke_getVariableGenes(assayData(umi.qc)
## $norm_counts, : Only 21 spike-ins to be used in fitting, may result in poor
## fit.</code></pre>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-brennecke-1.png" alt="(\#fig:rm-conf-brennecke)Results of using the Brennecke method on the Blischak dataset" width="90%" />
<p class="caption">
(#fig:rm-conf-brennecke)Results of using the Brennecke method on the Blischak dataset
</p>
</div>
<p>In the figure above blue points are the ERCC spike-ins. The red curve is the fitted technical noise model and the dashed line is the 95% CI. Pink dots are the genes with significant biological variability after multiple-testing correction. Since our dataset is relatively homogeneous only 148 genes are identified as significantly variable.</p>
</div>
<div id="remove-unwanted-variation" class="section level2">
<h2><span class="header-section-number">14.3</span> Remove Unwanted Variation</h2>
<p>Factors contributing to technical noise frequently appear as “batch effects” where cells processed on different days or by different technicians systematically vary from one another. Removing technical noise and correcting for batch effects can frequently be performed using the same tool or slight variants on it. We will be considering the <a href="http://bioconductor.org/packages/release/bioc/html/RUVSeq.html">Remove Unwanted Variation (RUV)</a> method which uses <a href="https://en.wikipedia.org/wiki/Singular_value_decomposition">singular value decomposition</a> (similar to PCA) on subsets of the dataset which should be invariant (e.g. <a href="https://www.thermofisher.com/order/catalog/product/4456740">ERCC spike-ins</a>). Then the method removes the identified unwanted factors.</p>
</div>
<div id="effectiveness-1" class="section level2">
<h2><span class="header-section-number">14.4</span> Effectiveness 1</h2>
<p>We evaluate the effectiveness of the normalization by inspecting the PCA plot where shape corresponds the technical replicate and colour corresponds to different biological samples (individuals from whom the iPSC lines where derived). Separation of biological samples and interspersed batches indicates that technical variation has been removed.</p>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-pca-rle-1.png" alt="(\#fig:rm-conf-pca-rle)PCA plot of the blischak data after RLE normalisation" width="90%" />
<p class="caption">
(#fig:rm-conf-pca-rle)PCA plot of the blischak data after RLE normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-pca-rle-ruv-1.png" alt="(\#fig:rm-conf-pca-rle-ruv)PCA plot of the blischak data after RLE and RUV normalisations" width="90%" />
<p class="caption">
(#fig:rm-conf-pca-rle-ruv)PCA plot of the blischak data after RLE and RUV normalisations
</p>
</div>
</div>
<div id="effectiveness-2" class="section level2">
<h2><span class="header-section-number">14.5</span> Effectiveness 2</h2>
<p>We can also examine the relative log expression (RLE) across cells to confirm technical noise has been removed from the dataset.</p>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-rle-comp-1.png" alt="(\#fig:rm-conf-rle-comp)Comparison of the relative log expression of the blischak data before and after the RUV normalisation" width="90%" />
<p class="caption">
(#fig:rm-conf-rle-comp)Comparison of the relative log expression of the blischak data before and after the RUV normalisation
</p>
</div>
</div>
<div id="exercise-4" class="section level2">
<h2><span class="header-section-number">14.6</span> Exercise</h2>
<p>Perform the same analysis with read counts of the Blischak data. Use <code>blischak/reads.rds</code> file to load the reads SCESet object. Once you have finished please compare your results to ours (next chapter). Additionally, experiment with other combinations of normalizations and compare the results.</p>
<!--chapter:end:13-remove-conf.Rmd-->
</div>
</div>
<div id="remove-confounders-using-controls-reads" class="section level1">
<h1><span class="header-section-number">15</span> Remove confounders using controls (Reads)</h1>
<pre><code>## Warning in scRNA.seq.funcs::Brennecke_getVariableGenes(assayData(reads.qc)
## $norm_counts, : Only 17 spike-ins to be used in fitting, may result in poor
## fit.</code></pre>
<pre><code>## Warning in xy.coords(x, y, xlabel, ylabel, log): 7 x values &lt;= 0 omitted
## from logarithmic plot</code></pre>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-brennecke-reads-1.png" alt="(\#fig:rm-conf-brennecke-reads)Results of using the Brennecke method on the Blischak dataset" width="90%" />
<p class="caption">
(#fig:rm-conf-brennecke-reads)Results of using the Brennecke method on the Blischak dataset
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-pca-rle-reads-1.png" alt="(\#fig:rm-conf-pca-rle-reads)PCA plot of the blischak data after RLE normalisation" width="90%" />
<p class="caption">
(#fig:rm-conf-pca-rle-reads)PCA plot of the blischak data after RLE normalisation
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-pca-rle-ruv-reads-1.png" alt="(\#fig:rm-conf-pca-rle-ruv-reads)PCA plot of the blischak data after RLE and RUV normalisations" width="90%" />
<p class="caption">
(#fig:rm-conf-pca-rle-ruv-reads)PCA plot of the blischak data after RLE and RUV normalisations
</p>
</div>
<div class="figure" style="text-align: center">
<img src="scRNA-seq-course_files/figure-html/rm-conf-rle-comp-reads-1.png" alt="(\#fig:rm-conf-rle-comp-reads)Comparison of the relative log expression of the blischak data before and after the RUV normalisation" width="90%" />
<p class="caption">
(#fig:rm-conf-rle-comp-reads)Comparison of the relative log expression of the blischak data before and after the RUV normalisation
</p>
</div>
<!--chapter:end:14-remove-conf-reads.Rmd-->
<div id="refs" class="references">

</div>
</div>
<!--bookdown:body:end-->
            </section>

          </div>
        </div>
      </div>
<!--bookdown:link_prev-->
<!--bookdown:link_next-->

<!--bookdown:config-->


</body>

</html>
